// @blocknote/core/dist/blocknote.js
var Fe = Object.defineProperty;
var Ve = (o, e, t) => e in o ? Fe(o, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : o[e] = t;
var k = (o, e, t) => Ve(o, typeof e != "symbol" ? e + "" : e, t);
import { Slice as F, Fragment as M, DOMSerializer as me, DOMParser as Ue, Node as ze } from "prosemirror-model";
import { ReplaceStep as Re, ReplaceAroundStep as j } from "prosemirror-transform";
import { n as q, i as V, g as E, b as Z, a as C, t as Ge, U, q as D, r as z, d as We, s as je, u as qe, v, w as B, x as ke, y as Ke, z as be, A as K } from "./blockToNode-BNoNIXU7.js";
import { B as gs, G as Bs, C as ys, D as Cs, N as Ss, E as xs, O as Es, F as Ps, H as Ts, I as ws, L as Ms, J as vs, M as Is, K as As } from "./blockToNode-BNoNIXU7.js";
import { ak as G, al as Je, am as Ye, an as Qe, aj as P, a as Xe, ao as Ze, ap as et, a5 as tt, a8 as J, aq as ot, ar as nt, a6 as st, as as ee, a9 as ge, at as rt } from "./defaultBlocks-ZzGbYgQn.js";
import { aw as Ls, av as _s, E as Ds, F as Os, r as $s, N as Hs, a4 as Fs, ad as Vs, aG as Us, ax as zs, b as Rs, d as Gs, e as Ws, a0 as js, aM as qs, au as Ks, c as Js, f as Ys, ag as Qs, ah as Xs, aB as Zs, x as er, y as tr, A as or, z as nr, g as sr, h as rr, T as ir, j as ar, k as cr, l as lr, n as dr, o as ur, q as pr, s as fr, w as hr, aC as mr, aH as kr, B as br, C as gr, H as Br, I as yr, J as Cr, K as Sr, aE as xr, aI as Er, M as Pr, D as Tr, G as wr, S as Mr, O as vr, Q as Ir, W as Ar, U as Nr, _ as Lr, Z as _r, a2 as Dr, Y as Or, X as $r, R as Hr, $ as Fr, m as Vr, aN as Ur, ac as zr, aK as Rr, az as Gr, ab as Wr, af as jr, i as qr, ay as Kr, aD as Jr, ae as Yr, a7 as Qr, t as Xr, u as Zr, v as ei, aa as ti, aJ as oi, ai as ni, aL as si, a1 as ri, V as ii, p as ai, a3 as ci, aF as li, L as di, aO as ui, P as pi, aA as fi } from "./defaultBlocks-ZzGbYgQn.js";
import { l as it, m as Be, n as at, o as ct, p as lt, d as Y, F as dt, Y as ut, a as pt, b as ft, S as ht, B as ye, c as mt, L as kt, N as bt, P as gt, h as Bt, k as yt, H as Ct, j as St, f as xt, V as Et, e as Pt } from "./TrailingNode-GzE59m_7.js";
import { D as mi, g as ki, T as bi, i as gi, s as Bi, t as yi } from "./TrailingNode-GzE59m_7.js";
import { s as Tt, B as wt } from "./BlockNoteSchema-ooiKsd5B.js";
import { C as Si, b as xi, c as Ei, a as Pi, g as Ti, u as wi, w as Mi } from "./BlockNoteSchema-ooiKsd5B.js";
import { Node as O, Extension as T, mergeAttributes as Mt, Mark as te, extensions as N, isNodeSelection as vt, posToDOMRect as It, selectionToInsertionEnd as At, findParentNodeClosestToPos as Nt, getSchema as Lt, createDocument as _t, Editor as Dt } from "@tiptap/core";
import { E as Ce } from "./EventEmitter-CjSwpTbz.js";
import { Fragment as R, Slice as W } from "@tiptap/pm/model";
import { e as Ot } from "./en-njEqD7AG.js";
import { inputRules as $t, InputRule as Ht } from "@handlewithcare/prosemirror-inputrules";
import { keymap as Ft } from "@tiptap/pm/keymap";
import { c as Vt, o as Ut } from "./BlockNoteExtension-C2X7LW-V.js";
import { a as Ii } from "./BlockNoteExtension-C2X7LW-V.js";
import { Gapcursor as zt } from "@tiptap/extensions/gap-cursor";
import { Link as Rt } from "@tiptap/extension-link";
import { Text as Gt } from "@tiptap/extension-text";
import { NodeSelection as $, TextSelection as S, Plugin as oe } from "prosemirror-state";
import { CellSelection as ne, TableMap as re } from "prosemirror-tables";
import { S as Wt } from "./ShowSelection-B0ch3unP.js";
import jt from "remark-gfm";
import qt from "remark-parse";
import Kt, { defaultHandlers as ie } from "remark-rehype";
import Jt from "rehype-stringify";
import { unified as Yt } from "unified";
import { TextSelection as Qt } from "@tiptap/pm/state";
function Xt(o, e) {
  const t = [
    {
      tag: `[data-inline-content-type="${o.type}"]`,
      contentElement: (n) => {
        const s = n;
        return s.matches("[data-editable]") ? s : s.querySelector("[data-editable]") || s;
      }
    }
  ];
  return e && t.push({
    tag: "*",
    getAttrs(n) {
      if (typeof n == "string")
        return !1;
      const s = e == null ? void 0 : e(n);
      return s === void 0 ? !1 : s;
    }
  }), t;
}
function ps(o, e) {
  var n;
  const t = O.create({
    name: o.type,
    inline: !0,
    group: "inline",
    draggable: (n = e.meta) == null ? void 0 : n.draggable,
    selectable: o.content === "styled",
    atom: o.content === "none",
    content: o.content === "styled" ? "inline*" : "",
    addAttributes() {
      return Ye(o.propSchema);
    },
    addKeyboardShortcuts() {
      return Je(o);
    },
    parseHTML() {
      return Xt(
        o,
        e.parse
      );
    },
    renderHTML({ node: s }) {
      const r = this.options.editor, i = e.render.call(
        { renderType: "dom", props: void 0 },
        q(
          s,
          r.schema.inlineContentSchema,
          r.schema.styleSchema
        ),
        // TODO: fix cast
        () => {
        },
        r
      );
      return G(
        i,
        o.type,
        s.attrs,
        o.propSchema
      );
    },
    addNodeView() {
      return (s) => {
        const { node: r, getPos: i } = s, l = this.options.editor, a = e.render.call(
          { renderType: "nodeView", props: s },
          q(
            r,
            l.schema.inlineContentSchema,
            l.schema.styleSchema
          ),
          // TODO: fix cast
          (c) => {
            const d = V([c], l.pmSchema), u = i();
            u && l.transact(
              (p) => p.replaceWith(u, u + r.nodeSize, d)
            );
          },
          l
        );
        return G(
          a,
          o.type,
          r.attrs,
          o.propSchema
        );
      };
    }
  });
  return Qe(
    t,
    o.propSchema,
    {
      ...e,
      toExternalHTML: e.toExternalHTML,
      render(s, r, i) {
        const l = e.render(
          s,
          r,
          i
        );
        return G(
          l,
          o.type,
          s.props,
          o.propSchema
        );
      }
    }
  );
}
function Zt(o, e, t, n = "before") {
  const s = typeof t == "string" ? t : t.id, r = E(o), i = e.map(
    (d) => Z(d, r)
  ), l = P(s, o.doc);
  if (!l)
    throw new Error(`Block with ID ${s} not found`);
  let a = l.posBeforeNode;
  return n === "after" && (a += l.node.nodeSize), o.step(
    new Re(a, a, new F(M.from(i), 0, 0))
  ), i.map(
    (d) => C(d, r)
  );
}
function Q(o) {
  if (!o || o.type.name !== "column")
    throw new Error("Invalid columnPos: does not point to column node.");
  const e = o.firstChild;
  if (!e)
    throw new Error("Invalid column: does not have child node.");
  const t = e.firstChild;
  if (!t)
    throw new Error("Invalid blockContainer: does not have child node.");
  return o.childCount === 1 && e.childCount === 1 && t.type.name === "paragraph" && t.content.content.length === 0;
}
function eo(o, e) {
  const t = o.doc.resolve(e), n = t.nodeAfter;
  if (!n || n.type.name !== "columnList")
    throw new Error(
      "Invalid columnListPos: does not point to columnList node."
    );
  for (let s = n.childCount - 1; s >= 0; s--) {
    const r = o.doc.resolve(t.pos + 1).posAtIndex(s), l = o.doc.resolve(r).nodeAfter;
    if (!l || l.type.name !== "column")
      throw new Error("Invalid columnPos: does not point to column node.");
    Q(l) && o.delete(r, r + l.nodeSize);
  }
}
function H(o, e) {
  eo(o, e);
  const n = o.doc.resolve(e).nodeAfter;
  if (!n || n.type.name !== "columnList")
    throw new Error(
      "Invalid columnListPos: does not point to columnList node."
    );
  if (n.childCount > 2)
    return;
  if (n.childCount < 2)
    throw new Error("Invalid columnList: contains fewer than two children.");
  const s = e + 1, i = o.doc.resolve(s).nodeAfter, l = e + n.nodeSize - 1, c = o.doc.resolve(l).nodeBefore;
  if (!i || !c)
    throw new Error("Invalid columnList: does not contain children.");
  const d = Q(i), u = Q(c);
  if (d && u) {
    o.delete(e, e + n.nodeSize);
    return;
  }
  if (d) {
    o.step(
      new j(
        // Replaces `columnList`.
        e,
        e + n.nodeSize,
        // Replaces with content of last `column`.
        l - c.nodeSize + 1,
        l - 1,
        // Doesn't append anything.
        F.empty,
        0,
        !1
      )
    );
    return;
  }
  if (u) {
    o.step(
      new j(
        // Replaces `columnList`.
        e,
        e + n.nodeSize,
        // Replaces with content of first `column`.
        s + 1,
        s + i.nodeSize - 1,
        // Doesn't append anything.
        F.empty,
        0,
        !1
      )
    );
    return;
  }
}
function ae(o, e, t) {
  const n = E(o), s = t.map(
    (u) => Z(u, n)
  ), r = new Set(
    e.map(
      (u) => typeof u == "string" ? u : u.id
    )
  ), i = [], l = /* @__PURE__ */ new Set(), a = typeof e[0] == "string" ? e[0] : e[0].id;
  let c = 0;
  if (o.doc.descendants((u, p) => {
    if (r.size === 0)
      return !1;
    if (!u.type.isInGroup("bnBlock") || !r.has(u.attrs.id))
      return !0;
    if (i.push(C(u, n)), r.delete(u.attrs.id), t.length > 0 && u.attrs.id === a) {
      const b = o.doc.nodeSize;
      o.insert(p, s);
      const g = o.doc.nodeSize;
      c += b - g;
    }
    const h = o.doc.nodeSize, f = o.doc.resolve(p - c);
    f.node().type.name === "column" ? l.add(f.before(-1)) : f.node().type.name === "columnList" && l.add(f.before()), f.node().type.name === "blockGroup" && f.node(f.depth - 1).type.name !== "doc" && f.node().childCount === 1 ? o.delete(f.before(), f.after()) : o.delete(p - c, p - c + u.nodeSize);
    const m = o.doc.nodeSize;
    return c += h - m, !1;
  }), r.size > 0) {
    const u = [...r].join(`
`);
    throw Error(
      "Blocks with the following IDs could not be found in the editor: " + u
    );
  }
  return l.forEach((u) => H(o, u)), { insertedBlocks: s.map(
    (u) => C(u, n)
  ), removedBlocks: i };
}
function to(o, e, t, n, s) {
  let r;
  if (e)
    if (typeof e == "string")
      r = V([e], o.pmSchema, n);
    else if (Array.isArray(e))
      r = V(e, o.pmSchema, n);
    else if (e.type === "tableContent")
      r = Ge(e, o.pmSchema);
    else
      throw new U(e.type);
  else throw new Error("blockContent is required");
  const l = ((s == null ? void 0 : s.document) ?? document).createDocumentFragment();
  for (const a of r)
    if (a.type.name !== "text" && o.schema.inlineContentSchema[a.type.name]) {
      const c = o.schema.inlineContentSpecs[a.type.name].implementation;
      if (c) {
        const d = q(
          a,
          o.schema.inlineContentSchema,
          o.schema.styleSchema
        ), u = c.render.call(
          {
            renderType: "dom",
            props: void 0
          },
          d,
          () => {
          },
          o
        );
        if (u) {
          if (l.appendChild(u.dom), u.contentDOM) {
            const p = t.serializeFragment(
              a.content,
              s
            );
            u.contentDOM.dataset.editable = "", u.contentDOM.appendChild(p);
          }
          continue;
        }
      }
    } else if (a.type.name === "text") {
      let c = document.createTextNode(
        a.textContent
      );
      for (const d of a.marks.toReversed())
        if (d.type.name in o.schema.styleSpecs) {
          const u = o.schema.styleSpecs[d.type.name].implementation.render(d.attrs.stringValue, o);
          u.contentDOM.appendChild(c), c = u.dom;
        } else {
          const u = d.type.spec.toDOM(d, !0), p = me.renderSpec(document, u);
          p.contentDOM.appendChild(c), c = p.dom;
        }
      l.appendChild(c);
    } else {
      const c = t.serializeFragment(
        M.from([a]),
        s
      );
      l.appendChild(c);
    }
  return l;
}
function oo(o, e, t, n) {
  var u, p, h, f, m;
  const s = o.pmSchema.nodes.blockContainer, r = e.props || {};
  for (const [b, g] of Object.entries(
    o.schema.blockSchema[e.type].propSchema
  ))
    !(b in r) && g.default !== void 0 && (r[b] = g.default);
  const i = e.children || [], a = o.blockImplementations[e.type].implementation.render.call(
    {
      renderType: "dom",
      props: void 0
    },
    { ...e, props: r, children: i },
    o
  );
  if (a.contentDOM && e.content) {
    const b = to(
      o,
      e.content,
      // TODO
      t,
      e.type,
      n
    );
    a.contentDOM.appendChild(b);
  }
  if (o.pmSchema.nodes[e.type].isInGroup("bnBlock")) {
    if (e.children && e.children.length > 0) {
      const b = Se(
        o,
        e.children,
        t,
        n
      );
      (u = a.contentDOM) == null || u.append(b);
    }
    return a.dom;
  }
  const d = (h = (p = s.spec) == null ? void 0 : p.toDOM) == null ? void 0 : h.call(
    p,
    s.create({
      id: e.id,
      ...r
    })
  );
  return (f = d.contentDOM) == null || f.appendChild(a.dom), e.children && e.children.length > 0 && ((m = d.contentDOM) == null || m.appendChild(
    xe(o, e.children, t, n)
  )), d.dom;
}
function Se(o, e, t, n) {
  const r = ((n == null ? void 0 : n.document) ?? document).createDocumentFragment();
  for (const i of e) {
    const l = oo(o, i, t, n);
    r.appendChild(l);
  }
  return r;
}
const xe = (o, e, t, n) => {
  var l;
  const s = o.pmSchema.nodes.blockGroup, r = s.spec.toDOM(s.create({})), i = Se(o, e, t, n);
  return (l = r.contentDOM) == null || l.appendChild(i), r.dom;
}, no = (o) => (o.querySelectorAll(
  '[data-content-type="numberedListItem"]'
).forEach((t) => {
  var s, r;
  const n = (r = (s = t.closest(".bn-block-outer")) == null ? void 0 : s.previousElementSibling) == null ? void 0 : r.querySelector(
    '[data-content-type="numberedListItem"]'
  );
  if (!n)
    t.setAttribute(
      "data-index",
      t.getAttribute("data-start") || "1"
    );
  else {
    const i = n.getAttribute("data-index");
    t.setAttribute(
      "data-index",
      (parseInt(i || "0") + 1).toString()
    );
  }
}), o), so = (o) => (o.querySelectorAll(
  '[data-content-type="checkListItem"] input'
).forEach((t) => {
  t.disabled = !0;
}), o), ro = (o) => (o.querySelectorAll(
  '.bn-toggle-wrapper[data-show-children="false"]'
).forEach((t) => {
  t.setAttribute("data-show-children", "true");
}), o), io = (o) => (o.querySelectorAll('[data-content-type="table"] table').forEach((t) => {
  t.setAttribute(
    "style",
    `--default-cell-min-width: ${Xe}px;`
  ), t.setAttribute("data-show-children", "true");
}), o), ao = (o) => (o.querySelectorAll('[data-content-type="table"] table').forEach((t) => {
  var r;
  const n = document.createElement("div");
  n.className = "tableWrapper";
  const s = document.createElement("div");
  s.className = "tableWrapper-inner", n.appendChild(s), (r = t.parentElement) == null || r.appendChild(n), n.appendChild(t);
}), o), co = (o) => (o.querySelectorAll(
  ".bn-inline-content:empty"
).forEach((t) => {
  const n = document.createElement("span");
  n.className = "ProseMirror-trailingBreak", n.setAttribute("style", "display: inline-block;"), t.appendChild(n);
}), o), lo = (o, e) => {
  const t = me.fromSchema(o), n = [
    no,
    so,
    ro,
    io,
    ao,
    co
  ];
  return {
    serializeBlocks: (s, r) => {
      let i = xe(
        e,
        s,
        t,
        r
      );
      for (const l of n)
        i = l(i);
      return i.outerHTML;
    }
  };
};
function uo(o) {
  return o.transact((e) => {
    const t = D(e.doc, e.selection.anchor);
    if (e.selection instanceof ne)
      return {
        type: "cell",
        anchorBlockId: t.node.attrs.id,
        anchorCellOffset: e.selection.$anchorCell.pos - t.posBeforeNode,
        headCellOffset: e.selection.$headCell.pos - t.posBeforeNode
      };
    if (e.selection instanceof $)
      return {
        type: "node",
        anchorBlockId: t.node.attrs.id
      };
    {
      const n = D(e.doc, e.selection.head);
      return {
        type: "text",
        anchorBlockId: t.node.attrs.id,
        headBlockId: n.node.attrs.id,
        anchorOffset: e.selection.anchor - t.posBeforeNode,
        headOffset: e.selection.head - n.posBeforeNode
      };
    }
  });
}
function po(o, e) {
  var s, r;
  const t = (s = P(e.anchorBlockId, o.doc)) == null ? void 0 : s.posBeforeNode;
  if (t === void 0)
    throw new Error(
      `Could not find block with ID ${e.anchorBlockId} to update selection`
    );
  let n;
  if (e.type === "cell")
    n = ne.create(
      o.doc,
      t + e.anchorCellOffset,
      t + e.headCellOffset
    );
  else if (e.type === "node")
    n = $.create(o.doc, t + 1);
  else {
    const i = (r = P(e.headBlockId, o.doc)) == null ? void 0 : r.posBeforeNode;
    if (i === void 0)
      throw new Error(
        `Could not find block with ID ${e.headBlockId} to update selection`
      );
    n = S.create(
      o.doc,
      t + e.anchorOffset,
      i + e.headOffset
    );
  }
  o.setSelection(n);
}
function X(o) {
  return o.map((e) => e.type === "columnList" ? e.children.map((t) => X(t.children)).flat() : {
    ...e,
    children: X(e.children)
  }).flat();
}
function Ee(o, e, t) {
  o.transact((n) => {
    var i;
    const s = ((i = o.getSelection()) == null ? void 0 : i.blocks) || [
      o.getTextCursorPosition().block
    ], r = uo(o);
    o.removeBlocks(s), o.insertBlocks(X(s), e, t), po(n, r);
  });
}
function Pe(o) {
  return !o || o.type !== "columnList";
}
function Te(o, e, t) {
  let n, s;
  if (e ? e.children.length > 0 ? (n = e.children[e.children.length - 1], s = "after") : (n = e, s = "before") : t && (n = t, s = "before"), !n || !s)
    return;
  const r = o.getParentBlock(n);
  return Pe(r) ? { referenceBlock: n, placement: s } : Te(
    o,
    s === "after" ? n : o.getPrevBlock(n),
    r
  );
}
function we(o, e, t) {
  let n, s;
  if (e ? e.children.length > 0 ? (n = e.children[0], s = "before") : (n = e, s = "after") : t && (n = t, s = "after"), !n || !s)
    return;
  const r = o.getParentBlock(n);
  return Pe(r) ? { referenceBlock: n, placement: s } : we(
    o,
    s === "before" ? n : o.getNextBlock(n),
    r
  );
}
function fo(o) {
  o.transact(() => {
    const e = o.getSelection(), t = (e == null ? void 0 : e.blocks[0]) || o.getTextCursorPosition().block, n = Te(
      o,
      o.getPrevBlock(t),
      o.getParentBlock(t)
    );
    n && Ee(
      o,
      n.referenceBlock,
      n.placement
    );
  });
}
function ho(o) {
  o.transact(() => {
    const e = o.getSelection(), t = (e == null ? void 0 : e.blocks[(e == null ? void 0 : e.blocks.length) - 1]) || o.getTextCursorPosition().block, n = we(
      o,
      o.getNextBlock(t),
      o.getParentBlock(t)
    );
    n && Ee(
      o,
      n.referenceBlock,
      n.placement
    );
  });
}
function mo(o, e, t) {
  const { $from: n, $to: s } = o.selection, r = n.blockRange(
    s,
    (f) => f.childCount > 0 && (f.type.name === "blockGroup" || f.type.name === "column")
    // change necessary to not look at first item child type
  );
  if (!r)
    return !1;
  const i = r.startIndex;
  if (i === 0)
    return !1;
  const a = r.parent.child(i - 1);
  if (a.type !== e)
    return !1;
  const c = a.lastChild && a.lastChild.type === t, d = M.from(c ? e.create() : null), u = new F(
    M.from(
      e.create(null, M.from(t.create(null, d)))
      // change necessary to create "groupType" instead of parent.type
    ),
    c ? 3 : 1,
    0
  ), p = r.start, h = r.end;
  return o.step(
    new j(
      p - (c ? 3 : 1),
      h,
      p,
      h,
      u,
      1,
      !0
    )
  ).scrollIntoView(), !0;
}
function Me(o) {
  return o.transact((e) => mo(
    e,
    o.pmSchema.nodes.blockContainer,
    o.pmSchema.nodes.blockGroup
  ));
}
function ko(o) {
  o._tiptapEditor.commands.liftListItem("blockContainer");
}
function bo(o) {
  return o.transact((e) => {
    const { bnBlock: t } = z(e);
    return e.doc.resolve(t.beforePos).nodeBefore !== null;
  });
}
function go(o) {
  return o.transact((e) => {
    const { bnBlock: t } = z(e);
    return e.doc.resolve(t.beforePos).depth > 1;
  });
}
function Bo(o, e) {
  const t = typeof e == "string" ? e : e.id, n = E(o), s = P(t, o);
  if (s)
    return C(s.node, n);
}
function yo(o, e) {
  const t = typeof e == "string" ? e : e.id, n = P(t, o), s = E(o);
  if (!n)
    return;
  const i = o.resolve(n.posBeforeNode).nodeBefore;
  if (i)
    return C(i, s);
}
function Co(o, e) {
  const t = typeof e == "string" ? e : e.id, n = P(t, o), s = E(o);
  if (!n)
    return;
  const i = o.resolve(
    n.posBeforeNode + n.node.nodeSize
  ).nodeAfter;
  if (i)
    return C(i, s);
}
function So(o, e) {
  const t = typeof e == "string" ? e : e.id, n = E(o), s = P(t, o);
  if (!s)
    return;
  const r = o.resolve(s.posBeforeNode), i = r.node(), l = r.node(-1), a = l.type.name !== "doc" ? i.type.name === "blockGroup" ? l : i : void 0;
  if (a)
    return C(a, n);
}
class xo {
  constructor(e) {
    this.editor = e;
  }
  /**
   * Gets a snapshot of all top-level (non-nested) blocks in the editor.
   * @returns A snapshot of all top-level (non-nested) blocks in the editor.
   */
  get document() {
    return this.editor.transact((e) => We(e.doc, this.editor.pmSchema));
  }
  /**
   * Gets a snapshot of an existing block from the editor.
   * @param blockIdentifier The identifier of an existing block that should be
   * retrieved.
   * @returns The block that matches the identifier, or `undefined` if no
   * matching block was found.
   */
  getBlock(e) {
    return this.editor.transact((t) => Bo(t.doc, e));
  }
  /**
   * Gets a snapshot of the previous sibling of an existing block from the
   * editor.
   * @param blockIdentifier The identifier of an existing block for which the
   * previous sibling should be retrieved.
   * @returns The previous sibling of the block that matches the identifier.
   * `undefined` if no matching block was found, or it's the first child/block
   * in the document.
   */
  getPrevBlock(e) {
    return this.editor.transact((t) => yo(t.doc, e));
  }
  /**
   * Gets a snapshot of the next sibling of an existing block from the editor.
   * @param blockIdentifier The identifier of an existing block for which the
   * next sibling should be retrieved.
   * @returns The next sibling of the block that matches the identifier.
   * `undefined` if no matching block was found, or it's the last child/block in
   * the document.
   */
  getNextBlock(e) {
    return this.editor.transact((t) => Co(t.doc, e));
  }
  /**
   * Gets a snapshot of the parent of an existing block from the editor.
   * @param blockIdentifier The identifier of an existing block for which the
   * parent should be retrieved.
   * @returns The parent of the block that matches the identifier. `undefined`
   * if no matching block was found, or the block isn't nested.
   */
  getParentBlock(e) {
    return this.editor.transact(
      (t) => So(t.doc, e)
    );
  }
  /**
   * Traverses all blocks in the editor depth-first, and executes a callback for each.
   * @param callback The callback to execute for each block. Returning `false` stops the traversal.
   * @param reverse Whether the blocks should be traversed in reverse order.
   */
  forEachBlock(e, t = !1) {
    const n = this.document.slice();
    t && n.reverse();
    function s(r) {
      for (const i of r) {
        if (e(i) === !1)
          return !1;
        const l = t ? i.children.slice().reverse() : i.children;
        if (!s(l))
          return !1;
      }
      return !0;
    }
    s(n);
  }
  /**
   * Inserts new blocks into the editor. If a block's `id` is undefined, BlockNote generates one automatically. Throws an
   * error if the reference block could not be found.
   * @param blocksToInsert An array of partial blocks that should be inserted.
   * @param referenceBlock An identifier for an existing block, at which the new blocks should be inserted.
   * @param placement Whether the blocks should be inserted just before, just after, or nested inside the
   * `referenceBlock`.
   */
  insertBlocks(e, t, n = "before") {
    return this.editor.transact(
      (s) => Zt(s, e, t, n)
    );
  }
  /**
   * Updates an existing block in the editor. Since updatedBlock is a PartialBlock object, some fields might not be
   * defined. These undefined fields are kept as-is from the existing block. Throws an error if the block to update could
   * not be found.
   * @param blockToUpdate The block that should be updated.
   * @param update A partial block which defines how the existing block should be changed.
   */
  updateBlock(e, t) {
    return this.editor.transact((n) => Ze(n, e, t));
  }
  /**
   * Removes existing blocks from the editor. Throws an error if any of the blocks could not be found.
   * @param blocksToRemove An array of identifiers for existing blocks that should be removed.
   */
  removeBlocks(e) {
    return this.editor.transact(
      (t) => ae(t, e, []).removedBlocks
    );
  }
  /**
   * Replaces existing blocks in the editor with new blocks. If the blocks that should be removed are not adjacent or
   * are at different nesting levels, `blocksToInsert` will be inserted at the position of the first block in
   * `blocksToRemove`. Throws an error if any of the blocks to remove could not be found.
   * @param blocksToRemove An array of blocks that should be replaced.
   * @param blocksToInsert An array of partial blocks to replace the old ones with.
   */
  replaceBlocks(e, t) {
    return this.editor.transact(
      (n) => ae(n, e, t)
    );
  }
  /**
   * Checks if the block containing the text cursor can be nested.
   */
  canNestBlock() {
    return bo(this.editor);
  }
  /**
   * Nests the block containing the text cursor into the block above it.
   */
  nestBlock() {
    Me(this.editor);
  }
  /**
   * Checks if the block containing the text cursor is nested.
   */
  canUnnestBlock() {
    return go(this.editor);
  }
  /**
   * Lifts the block containing the text cursor out of its parent.
   */
  unnestBlock() {
    ko(this.editor);
  }
  /**
   * Moves the selected blocks up. If the previous block has children, moves
   * them to the end of its children. If there is no previous block, but the
   * current blocks share a common parent, moves them out of & before it.
   */
  moveBlocksUp() {
    return fo(this.editor);
  }
  /**
   * Moves the selected blocks down. If the next block has children, moves
   * them to the start of its children. If there is no next block, but the
   * current blocks share a common parent, moves them out of & after it.
   */
  moveBlocksDown() {
    return ho(this.editor);
  }
}
class Eo extends Ce {
  constructor(e) {
    super(), this.editor = e, e.on("create", () => {
      e._tiptapEditor.on(
        "update",
        ({ transaction: t, appendedTransactions: n }) => {
          this.emit("onChange", { editor: e, transaction: t, appendedTransactions: n });
        }
      ), e._tiptapEditor.on("selectionUpdate", ({ transaction: t }) => {
        this.emit("onSelectionChange", { editor: e, transaction: t });
      }), e._tiptapEditor.on("mount", () => {
        this.emit("onMount", { editor: e });
      }), e._tiptapEditor.on("unmount", () => {
        this.emit("onUnmount", { editor: e });
      });
    });
  }
  /**
   * Register a callback that will be called when the editor changes.
   */
  onChange(e, t = !0) {
    const n = ({
      transaction: s,
      appendedTransactions: r
    }) => {
      !t && ce(s) || e(this.editor, {
        getChanges() {
          return it(
            s,
            r
          );
        }
      });
    };
    return this.on("onChange", n), () => {
      this.off("onChange", n);
    };
  }
  /**
   * Register a callback that will be called when the selection changes.
   */
  onSelectionChange(e, t = !1) {
    const n = (s) => {
      !t && ce(s.transaction) || e(this.editor);
    };
    return this.on("onSelectionChange", n), () => {
      this.off("onSelectionChange", n);
    };
  }
  /**
   * Register a callback that will be called when the editor is mounted.
   */
  onMount(e) {
    return this.on("onMount", e), () => {
      this.off("onMount", e);
    };
  }
  /**
   * Register a callback that will be called when the editor is unmounted.
   */
  onUnmount(e) {
    return this.on("onUnmount", e), () => {
      this.off("onUnmount", e);
    };
  }
}
function ce(o) {
  return !!o.getMeta("y-sync$");
}
function Po(o) {
  return Array.prototype.indexOf.call(o.parentElement.childNodes, o);
}
function To(o) {
  return o.nodeType === 3 && !/\S/.test(o.nodeValue || "");
}
function wo(o) {
  o.querySelectorAll("li > ul, li > ol").forEach((e) => {
    const t = Po(e), n = e.parentElement, s = Array.from(n.childNodes).slice(
      t + 1
    );
    e.remove(), s.forEach((r) => {
      r.remove();
    }), n.insertAdjacentElement("afterend", e), s.reverse().forEach((r) => {
      if (To(r))
        return;
      const i = document.createElement("li");
      i.append(r), e.insertAdjacentElement("afterend", i);
    }), n.childNodes.length === 0 && n.remove();
  });
}
function Mo(o) {
  o.querySelectorAll("li + ul, li + ol").forEach((e) => {
    var r, i;
    const t = e.previousElementSibling, n = document.createElement("div");
    t.insertAdjacentElement("afterend", n), n.append(t);
    const s = document.createElement("div");
    for (s.setAttribute("data-node-type", "blockGroup"), n.append(s); ((r = n.nextElementSibling) == null ? void 0 : r.nodeName) === "UL" || ((i = n.nextElementSibling) == null ? void 0 : i.nodeName) === "OL"; )
      s.append(n.nextElementSibling);
  });
}
let le = null;
function vo() {
  return le || (le = document.implementation.createHTMLDocument("title"));
}
function Io(o) {
  if (typeof o == "string") {
    const e = vo().createElement("div");
    e.innerHTML = o, o = e;
  }
  return wo(o), Mo(o), o;
}
function Ao(o) {
  const e = o.ownerDocument.createTreeWalker(
    o,
    // NodeFilter.SHOW_COMMENT
    128
  );
  let t;
  for (; t = e.nextNode(); )
    if (/^\s*notionvc:/.test(t.nodeValue || ""))
      return !0;
  return !1;
}
function No(o) {
  const e = /* @__PURE__ */ new Set(["PRE", "CODE"]), t = o.ownerDocument.createTreeWalker(
    o,
    // NodeFilter.SHOW_TEXT
    4,
    {
      acceptNode(r) {
        let i = r.parentElement;
        for (; i && i !== o; ) {
          if (e.has(i.tagName))
            return 2;
          i = i.parentElement;
        }
        return 1;
      }
    }
  ), n = [];
  let s;
  for (; s = t.nextNode(); )
    n.push(s);
  for (const r of n)
    r.nodeValue && /[\r\n]/.test(r.nodeValue) && (r.nodeValue = r.nodeValue.replace(/[ \t\r\n\f]+/g, " "));
}
function Lo(o) {
  Ao(o) || No(o);
}
function ve(o, e) {
  const t = Io(o);
  Lo(t);
  const s = Ue.fromSchema(e).parse(t, {
    topNode: e.nodes.blockGroup.create()
  }), r = [];
  for (let i = 0; i < s.childCount; i++)
    r.push(C(s.child(i), e));
  return r;
}
function _o(o, e) {
  const t = e.value ? e.value : "", n = {};
  e.lang && (n["data-language"] = e.lang);
  let s = {
    type: "element",
    tagName: "code",
    properties: n,
    children: [{ type: "text", value: t }]
  };
  return e.meta && (s.data = { meta: e.meta }), o.patch(e, s), s = o.applyData(e, s), s = {
    type: "element",
    tagName: "pre",
    properties: {},
    children: [s]
  }, o.patch(e, s), s;
}
function Do(o, e) {
  var r;
  const t = String((e == null ? void 0 : e.url) || ""), n = e != null && e.title ? String(e.title) : void 0;
  let s = {
    type: "element",
    tagName: "video",
    properties: {
      src: t,
      "data-name": n,
      "data-url": t,
      controls: !0
    },
    children: []
  };
  return (r = o.patch) == null || r.call(o, e, s), s = o.applyData ? o.applyData(e, s) : s, s;
}
function Ie(o) {
  return Yt().use(qt).use(jt).use(Kt, {
    handlers: {
      ...ie,
      image: (t, n) => {
        const s = String((n == null ? void 0 : n.url) || "");
        return et(s) ? Do(t, n) : ie.image(t, n);
      },
      code: _o,
      blockquote: (t, n) => {
        const s = {
          type: "element",
          tagName: "blockquote",
          properties: {},
          // The only difference from the original is that we don't wrap the children with line endings
          children: t.wrap(t.all(n), !1)
        };
        return t.patch(n, s), t.applyData(n, s);
      }
    }
  }).use(Jt).processSync(o).value;
}
function Oo(o, e) {
  const t = Ie(o);
  return ve(t, e);
}
class $o {
  constructor(e) {
    this.editor = e;
  }
  /**
   * Exports blocks into a simplified HTML string. To better conform to HTML standards, children of blocks which aren't list
   * items are un-nested in the output HTML.
   *
   * @param blocks An array of blocks that should be serialized into HTML.
   * @returns The blocks, serialized as an HTML string.
   */
  blocksToHTMLLossy(e = this.editor.document) {
    return Be(
      this.editor.pmSchema,
      this.editor
    ).exportBlocks(e, {});
  }
  /**
   * Serializes blocks into an HTML string in the format that would normally be rendered by the editor.
   *
   * Use this method if you want to server-side render HTML (for example, a blog post that has been edited in BlockNote)
   * and serve it to users without loading the editor on the client (i.e.: displaying the blog post)
   *
   * @param blocks An array of blocks that should be serialized into HTML.
   * @returns The blocks, serialized as an HTML string.
   */
  blocksToFullHTML(e = this.editor.document) {
    return lo(
      this.editor.pmSchema,
      this.editor
    ).serializeBlocks(e, {});
  }
  /**
   * Parses blocks from an HTML string. Tries to create `Block` objects out of any HTML block-level elements, and
   * `InlineNode` objects from any HTML inline elements, though not all element types are recognized. If BlockNote
   * doesn't recognize an HTML element's tag, it will parse it as a paragraph or plain text.
   * @param html The HTML string to parse blocks from.
   * @returns The blocks parsed from the HTML string.
   */
  tryParseHTMLToBlocks(e) {
    return ve(e, this.editor.pmSchema);
  }
  /**
   * Serializes blocks into a Markdown string. The output is simplified as Markdown does not support all features of
   * BlockNote - children of blocks which aren't list items are un-nested and certain styles are removed.
   * @param blocks An array of blocks that should be serialized into Markdown.
   * @returns The blocks, serialized as a Markdown string.
   */
  blocksToMarkdownLossy(e = this.editor.document) {
    return at(e, this.editor.pmSchema, this.editor, {});
  }
  /**
   * Creates a list of blocks from a Markdown string. Tries to create `Block` and `InlineNode` objects based on
   * Markdown syntax, though not all symbols are recognized. If BlockNote doesn't recognize a symbol, it will parse it
   * as text.
   * @param markdown The Markdown string to parse blocks from.
   * @returns The blocks parsed from the Markdown string.
   */
  tryParseMarkdownToBlocks(e) {
    return Oo(e, this.editor.pmSchema);
  }
  /**
   * Paste HTML into the editor. Defaults to converting HTML to BlockNote HTML.
   * @param html The HTML to paste.
   * @param raw Whether to paste the HTML as is, or to convert it to BlockNote HTML.
   */
  pasteHTML(e, t = !1) {
    var s;
    let n = e;
    if (!t) {
      const r = this.tryParseHTMLToBlocks(e);
      n = this.blocksToFullHTML(r);
    }
    n && ((s = this.editor.prosemirrorView) == null || s.pasteHTML(n));
  }
  /**
   * Paste text into the editor. Defaults to interpreting text as markdown.
   * @param text The text to paste.
   */
  pasteText(e) {
    var t;
    return (t = this.editor.prosemirrorView) == null ? void 0 : t.pasteText(e);
  }
  /**
   * Paste markdown into the editor.
   * @param markdown The markdown to paste.
   */
  pasteMarkdown(e) {
    const t = Ie(e);
    return this.pasteHTML(t);
  }
}
const se = [
  "vscode-editor-data",
  "blocknote/html",
  "text/markdown",
  "text/html",
  "text/plain",
  "Files"
];
function Ho(o, e) {
  if (!o.startsWith(".") || !e.startsWith("."))
    throw new Error("The strings provided are not valid file extensions.");
  return o === e;
}
function Fo(o, e) {
  const t = o.split("/"), n = e.split("/");
  if (t.length !== 2)
    throw new Error(`The string ${o} is not a valid MIME type.`);
  if (n.length !== 2)
    throw new Error(`The string ${e} is not a valid MIME type.`);
  return t[1] === "*" || n[1] === "*" ? t[0] === n[0] : (t[0] === "*" || n[0] === "*" || t[0] === n[0]) && t[1] === n[1];
}
function de(o, e, t, n = "after") {
  let s;
  return Array.isArray(e.content) && e.content.length === 0 ? s = o.updateBlock(e, t).id : s = o.insertBlocks(
    [t],
    e,
    n
  )[0].id, s;
}
async function Ae(o, e) {
  var r;
  if (!e.uploadFile) {
    console.warn(
      "Attempted ot insert file, but uploadFile is not set in the BlockNote editor options"
    );
    return;
  }
  const t = "dataTransfer" in o ? o.dataTransfer : o.clipboardData;
  if (t === null)
    return;
  let n = null;
  for (const i of se)
    if (t.types.includes(i)) {
      n = i;
      break;
    }
  if (n !== "Files")
    return;
  const s = t.items;
  if (s) {
    o.preventDefault();
    for (let i = 0; i < s.length; i++) {
      let l = "file";
      for (const c of Object.values(e.schema.blockSpecs))
        for (const d of ((r = c.implementation.meta) == null ? void 0 : r.fileBlockAccept) || []) {
          const u = d.startsWith("."), p = s[i].getAsFile();
          if (p && (!u && p.type && Fo(s[i].type, d) || u && Ho(
            "." + p.name.split(".").pop(),
            d
          ))) {
            l = c.config.type;
            break;
          }
        }
      const a = s[i].getAsFile();
      if (a) {
        const c = {
          type: l,
          props: {
            name: a.name
          }
        };
        let d;
        if (o.type === "paste") {
          const h = e.getTextCursorPosition().block;
          d = de(e, h, c);
        } else if (o.type === "drop") {
          const h = {
            left: o.clientX,
            top: o.clientY
          }, f = e.prosemirrorView.posAtCoords(h);
          if (!f)
            return;
          d = e.transact((m) => {
            var y;
            const b = D(m.doc, f.pos), g = (y = e.domElement) == null ? void 0 : y.querySelector(
              `[data-id="${b.node.attrs.id}"]`
            ), x = g == null ? void 0 : g.getBoundingClientRect();
            return de(
              e,
              e.getBlock(b.node.attrs.id),
              c,
              x && (x.top + x.bottom) / 2 > h.top ? "before" : "after"
            );
          });
        } else
          return;
        const u = await e.uploadFile(a, d), p = typeof u == "string" ? {
          props: {
            url: u
          }
        } : { ...u };
        e.updateBlock(d, p);
      }
    }
  }
}
const Vo = (o) => T.create({
  name: "dropFile",
  addProseMirrorPlugins() {
    return [
      new oe({
        props: {
          handleDOMEvents: {
            drop(e, t) {
              if (!o.isEditable)
                return;
              let n = null;
              for (const s of se)
                if (t.dataTransfer.types.includes(s)) {
                  n = s;
                  break;
                }
              return n === null ? !0 : n === "Files" ? (Ae(t, o), !0) : !1;
            }
          }
        }
      })
    ];
  }
}), Uo = /(^|\n) {0,3}#{1,6} {1,8}[^\n]{1,64}\r?\n\r?\n\s{0,32}\S/, zo = /(_|__|\*|\*\*|~~|==|\+\+)(?!\s)(?:[^\s](?:.{0,62}[^\s])?|\S)(?=\1)/, Ro = /\[[^\]]{1,128}\]\(https?:\/\/\S{1,999}\)/, Go = /(?:\s|^)`(?!\s)(?:[^\s`](?:[^`]{0,46}[^\s`])?|[^\s`])`([^\w]|$)/, Wo = /(?:^|\n)\s{0,5}-\s{1}[^\n]+\n\s{0,15}-\s/, jo = /(?:^|\n)\s{0,5}\d+\.\s{1}[^\n]+\n\s{0,15}\d+\.\s/, qo = /\n{2} {0,3}-{2,48}\n{2}/, Ko = /(?:\n|^)(```|~~~|\$\$)(?!`|~)[^\s]{0,64} {0,64}[^\n]{0,64}\n[\s\S]{0,9999}?\s*\1 {0,64}(?:\n+|$)/, Jo = /(?:\n|^)(?!\s)\w[^\n]{0,64}\r?\n(-|=)\1{0,64}\n\n\s{0,64}(\w|$)/, Yo = /(?:^|(\r?\n\r?\n))( {0,3}>[^\n]{1,333}\n){1,999}($|(\r?\n))/, Qo = /^\s*\|(.+\|)+\s*$/m, Xo = /^\s*\|(\s*[-:]+[-:]\s*\|)+\s*$/m, Zo = /^\s*\|(.+\|)+\s*$/m, en = (o) => Uo.test(o) || zo.test(o) || Ro.test(o) || Go.test(o) || Wo.test(o) || jo.test(o) || qo.test(o) || Ko.test(o) || Jo.test(o) || Yo.test(o) || Qo.test(o) || Xo.test(o) || Zo.test(o);
async function tn(o, e) {
  const { schema: t } = e.state;
  if (!o.clipboardData)
    return !1;
  const n = o.clipboardData.getData("text/plain");
  if (!n)
    return !1;
  if (!t.nodes.codeBlock)
    return e.pasteText(n), !0;
  const s = o.clipboardData.getData("vscode-editor-data"), r = s ? JSON.parse(s) : void 0, i = r == null ? void 0 : r.mode;
  return i ? (e.pasteHTML(
    `<pre><code class="language-${i}">${n.replace(
      /\r\n?/g,
      `
`
    )}</code></pre>`
  ), !0) : !1;
}
function on({
  event: o,
  editor: e,
  prioritizeMarkdownOverHTML: t,
  plainTextAsMarkdown: n
}) {
  var l;
  if (e.transact(
    (a) => a.selection.$from.parent.type.spec.code && a.selection.$to.parent.type.spec.code
  )) {
    const a = (l = o.clipboardData) == null ? void 0 : l.getData("text/plain");
    if (a)
      return e.pasteText(a), !0;
  }
  let r;
  for (const a of se)
    if (o.clipboardData.types.includes(a)) {
      r = a;
      break;
    }
  if (!r)
    return !0;
  if (r === "vscode-editor-data")
    return tn(o, e.prosemirrorView), !0;
  if (r === "Files")
    return Ae(o, e), !0;
  const i = o.clipboardData.getData(r);
  if (r === "blocknote/html")
    return e.pasteHTML(i, !0), !0;
  if (r === "text/markdown")
    return e.pasteMarkdown(i), !0;
  if (t) {
    const a = o.clipboardData.getData("text/plain");
    if (en(a))
      return e.pasteMarkdown(a), !0;
  }
  return r === "text/html" ? (e.pasteHTML(i), !0) : n ? (e.pasteMarkdown(i), !0) : (e.pasteText(i), !0);
}
const nn = (o, e) => T.create({
  name: "pasteFromClipboard",
  addProseMirrorPlugins() {
    return [
      new oe({
        props: {
          handleDOMEvents: {
            paste(t, n) {
              if (n.preventDefault(), !!o.isEditable)
                return e({
                  event: n,
                  editor: o,
                  defaultPasteHandler: ({
                    prioritizeMarkdownOverHTML: s = !0,
                    plainTextAsMarkdown: r = !0
                  } = {}) => on({
                    event: n,
                    editor: o,
                    prioritizeMarkdownOverHTML: s,
                    plainTextAsMarkdown: r
                  })
                });
            }
          }
        }
      })
    ];
  }
});
function sn(o, e, t) {
  var l;
  let n = !1;
  const s = o.state.selection instanceof ne;
  if (!s) {
    const a = o.state.doc.slice(
      o.state.selection.from,
      o.state.selection.to,
      !1
    ).content, c = [];
    for (let d = 0; d < a.childCount; d++)
      c.push(a.child(d));
    n = c.find(
      (d) => d.type.isInGroup("bnBlock") || d.type.name === "blockGroup" || d.type.spec.group === "blockContent"
    ) === void 0, n && (e = a);
  }
  let r;
  const i = Be(
    o.state.schema,
    t
  );
  if (s) {
    ((l = e.firstChild) == null ? void 0 : l.type.name) === "table" && (e = e.firstChild.content);
    const a = je(
      e,
      t.schema.inlineContentSchema,
      t.schema.styleSchema
    );
    r = `<table>${i.exportInlineContent(
      a,
      {}
    )}</table>`;
  } else if (n) {
    const a = qe(
      e,
      t.schema.inlineContentSchema,
      t.schema.styleSchema
    );
    r = i.exportInlineContent(a, {});
  } else {
    const a = lt(e);
    r = i.exportBlocks(a, {});
  }
  return r;
}
function Ne(o, e) {
  "node" in o.state.selection && o.state.selection.node.type.spec.group === "blockContent" && e.transact(
    (i) => i.setSelection(
      new $(i.doc.resolve(o.state.selection.from - 1))
    )
  );
  const t = o.serializeForClipboard(
    o.state.selection.content()
  ).dom.innerHTML, n = o.state.selection.content().content, s = sn(
    o,
    n,
    e
  ), r = ct(s);
  return { clipboardHTML: t, externalHTML: s, markdown: r };
}
const ue = () => {
  const o = window.getSelection();
  if (!o || o.isCollapsed)
    return !0;
  let e = o.focusNode;
  for (; e; ) {
    if (e instanceof HTMLElement && e.getAttribute("contenteditable") === "false")
      return !0;
    e = e.parentElement;
  }
  return !1;
}, pe = (o, e, t) => {
  t.preventDefault(), t.clipboardData.clearData();
  const { clipboardHTML: n, externalHTML: s, markdown: r } = Ne(
    e,
    o
  );
  t.clipboardData.setData("blocknote/html", n), t.clipboardData.setData("text/html", s), t.clipboardData.setData("text/plain", r);
}, rn = (o) => T.create({
  name: "copyToClipboard",
  addProseMirrorPlugins() {
    return [
      new oe({
        props: {
          handleDOMEvents: {
            copy(e, t) {
              return ue() || pe(o, e, t), !0;
            },
            cut(e, t) {
              return ue() || (pe(o, e, t), e.editable && e.dispatch(e.state.tr.deleteSelection())), !0;
            },
            // This is for the use-case in which only a block without content
            // is selected, e.g. an image block, and dragged (not using the
            // drag handle).
            dragstart(e, t) {
              if (!("node" in e.state.selection) || e.state.selection.node.type.spec.group !== "blockContent")
                return;
              o.transact(
                (i) => i.setSelection(
                  new $(
                    i.doc.resolve(e.state.selection.from - 1)
                  )
                )
              ), t.preventDefault(), t.dataTransfer.clearData();
              const { clipboardHTML: n, externalHTML: s, markdown: r } = Ne(e, o);
              return t.dataTransfer.setData("blocknote/html", n), t.dataTransfer.setData("text/html", s), t.dataTransfer.setData("text/plain", r), !0;
            }
          }
        }
      })
    ];
  }
}), an = T.create({
  name: "blockBackgroundColor",
  addGlobalAttributes() {
    return [
      {
        types: ["tableCell", "tableHeader"],
        attributes: {
          backgroundColor: tt()
        }
      }
    ];
  }
}), cn = O.create({
  name: "hardBreak",
  inline: !0,
  group: "inline",
  selectable: !1,
  linebreakReplacement: !0,
  priority: 10,
  parseHTML() {
    return [{ tag: "br" }];
  },
  renderHTML({ HTMLAttributes: o }) {
    return ["br", Mt(this.options.HTMLAttributes, o)];
  },
  renderText() {
    return `
`;
  }
}), Le = (o, e) => {
  var l;
  const t = o.resolve(e), n = t.depth - 1, s = t.before(n), r = o.resolve(s).nodeAfter;
  return r ? (l = r.type.spec.group) != null && l.includes("bnBlock") ? v(
    o.resolve(s)
  ) : Le(o, s) : void 0;
}, _ = (o, e) => {
  const t = o.resolve(e), n = t.index();
  if (n === 0)
    return;
  const s = t.posAtIndex(n - 1);
  return v(
    o.resolve(s)
  );
}, L = (o, e) => {
  const t = o.resolve(e), n = t.index();
  if (n === t.node().childCount - 1)
    return;
  const s = t.posAtIndex(n + 1);
  return v(
    o.resolve(s)
  );
}, _e = (o, e) => {
  for (; e.childContainer; ) {
    const t = e.childContainer.node, n = o.resolve(e.childContainer.beforePos + 1).posAtIndex(t.childCount - 1);
    e = v(o.resolve(n));
  }
  return e;
}, ln = (o, e) => o.isBlockContainer && o.blockContent.node.type.spec.content === "inline*" && o.blockContent.node.childCount > 0 && e.isBlockContainer && e.blockContent.node.type.spec.content === "inline*", dn = (o, e, t, n) => {
  if (!n.isBlockContainer)
    throw new Error(
      `Attempted to merge block at position ${n.bnBlock.beforePos} into previous block at position ${t.bnBlock.beforePos}, but next block is not a block container`
    );
  if (n.childContainer) {
    const s = o.doc.resolve(
      n.childContainer.beforePos + 1
    ), r = o.doc.resolve(
      n.childContainer.afterPos - 1
    ), i = s.blockRange(r);
    if (e) {
      const l = o.doc.resolve(n.bnBlock.beforePos);
      o.tr.lift(i, l.depth);
    }
  }
  if (e) {
    if (!t.isBlockContainer)
      throw new Error(
        `Attempted to merge block at position ${n.bnBlock.beforePos} into previous block at position ${t.bnBlock.beforePos}, but previous block is not a block container`
      );
    e(
      o.tr.delete(
        t.blockContent.afterPos - 1,
        n.blockContent.beforePos + 1
      )
    );
  }
  return !0;
}, fe = (o) => ({
  state: e,
  dispatch: t
}) => {
  const n = e.doc.resolve(o), s = v(n), r = _(
    e.doc,
    s.bnBlock.beforePos
  );
  if (!r)
    return !1;
  const i = _e(
    e.doc,
    r
  );
  return ln(i, s) ? dn(e, t, i, s) : !1;
}, un = T.create({
  priority: 50,
  // TODO: The shortcuts need a refactor. Do we want to use a command priority
  //  design as there is now, or clump the logic into a single function?
  addKeyboardShortcuts() {
    const o = () => this.editor.commands.first(({ chain: n, commands: s }) => [
      // Deletes the selection if it's not empty.
      () => s.deleteSelection(),
      // Undoes an input rule if one was triggered in the last editor state change.
      () => s.undoInputRule(),
      // Reverts block content type to a paragraph if the selection is at the start of the block.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const l = r.selection.from === i.blockContent.beforePos + 1, a = i.blockContent.node.type.name === "paragraph";
        return l && !a ? s.command(
          nt(i.bnBlock.beforePos, {
            type: "paragraph",
            props: {}
          })
        ) : !1;
      }),
      // Removes a level of nesting if the block is indented if the selection is at the start of the block.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const { blockContent: l } = i;
        return r.selection.from === l.beforePos + 1 ? s.liftListItem("blockContainer") : !1;
      }),
      // Merges block with the previous one if it isn't indented, and the selection is at the start of the
      // block. The target block for merging must contain inline content.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const { bnBlock: l, blockContent: a } = i, c = _(
          r.doc,
          i.bnBlock.beforePos
        );
        if (!c || !c.isBlockContainer || c.blockContent.node.type.spec.content !== "inline*")
          return !1;
        const d = r.selection.from === a.beforePos + 1, u = r.selection.empty, p = l.beforePos;
        return d && u ? n().command(fe(p)).scrollIntoView().run() : !1;
      }),
      // If the previous block is a columnList, moves the current block to
      // the end of the last column in it.
      () => s.command(({ state: r, tr: i, dispatch: l }) => {
        const a = B(r);
        if (!a.isBlockContainer)
          return !1;
        const c = _(
          r.doc,
          a.bnBlock.beforePos
        );
        if (!c || c.isBlockContainer)
          return !1;
        if (l) {
          const d = c.bnBlock.afterPos - 1, u = i.doc.resolve(d - 1);
          return i.delete(
            a.bnBlock.beforePos,
            a.bnBlock.afterPos
          ), i.insert(u.pos, a.bnBlock.node), i.setSelection(
            S.near(i.doc.resolve(u.pos + 1))
          ), !0;
        }
        return !1;
      }),
      // If the block is the first in a column, moves it to the end of the
      // previous column. If there is no previous column, moves it above the
      // columnList.
      () => s.command(({ state: r, tr: i, dispatch: l }) => {
        const a = B(r);
        if (!a.isBlockContainer || !(i.selection.from === a.blockContent.beforePos + 1))
          return !1;
        const d = i.doc.resolve(a.bnBlock.beforePos);
        if (d.nodeBefore || d.node().type.name !== "column")
          return !1;
        const h = i.doc.resolve(a.bnBlock.beforePos), f = i.doc.resolve(h.before()), m = f.before();
        return l && (i.delete(
          a.bnBlock.beforePos,
          a.bnBlock.afterPos
        ), H(i, m), f.pos === m + 1 ? (i.insert(m, a.bnBlock.node), i.setSelection(
          S.near(i.doc.resolve(m))
        )) : (i.insert(f.pos - 1, a.bnBlock.node), i.setSelection(
          S.near(i.doc.resolve(f.pos))
        ))), !0;
      }),
      // Deletes the current block if it's an empty block with inline content,
      // and moves the selection to the previous block.
      () => s.command(({ state: r }) => {
        var a;
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        if (i.blockContent.node.childCount === 0 && i.blockContent.node.type.spec.content === "inline*") {
          const c = _(
            r.doc,
            i.bnBlock.beforePos
          );
          if (!c || !c.isBlockContainer)
            return !1;
          let d = n();
          if (i.childContainer && d.insertContentAt(
            i.bnBlock.afterPos,
            (a = i.childContainer) == null ? void 0 : a.node.content
          ), c.blockContent.node.type.spec.content === "tableRow+") {
            const m = i.bnBlock.beforePos - 1 - 1 - 1 - 1 - 1;
            d = d.setTextSelection(
              m
            );
          } else if (c.blockContent.node.type.spec.content === "")
            d = d.setNodeSelection(
              c.blockContent.beforePos
            );
          else {
            const u = c.blockContent.afterPos - 1;
            d = d.setTextSelection(u);
          }
          return d.deleteRange({
            from: i.bnBlock.beforePos,
            to: i.bnBlock.afterPos
          }).scrollIntoView().run();
        }
        return !1;
      }),
      // Deletes previous block if it contains no content and isn't a table,
      // when the selection is empty and at the start of the block. Moves the
      // current block into the deleted block's place.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const l = r.selection.from === i.blockContent.beforePos + 1, a = r.selection.empty, c = _(
          r.doc,
          i.bnBlock.beforePos
        );
        if (c && l && a) {
          const d = _e(
            r.doc,
            c
          );
          if (!d.isBlockContainer)
            return !1;
          if (d.blockContent.node.type.spec.content === "" || d.blockContent.node.type.spec.content === "inline*" && d.blockContent.node.childCount === 0)
            return n().cut(
              {
                from: i.bnBlock.beforePos,
                to: i.bnBlock.afterPos
              },
              d.bnBlock.afterPos
            ).deleteRange({
              from: d.bnBlock.beforePos,
              to: d.bnBlock.afterPos
            }).run();
        }
        return !1;
      })
    ]), e = () => this.editor.commands.first(({ chain: n, commands: s }) => [
      // Deletes the selection if it's not empty.
      () => s.deleteSelection(),
      // Deletes the first child block and un-nests its children, if the
      // selection is empty and at the end of the current block. If both the
      // parent and child blocks have inline content, the child block's
      // content is appended to the parent's. The child block's own children
      // are unindented before it's deleted.
      () => s.command(({ state: r }) => {
        var p;
        const i = B(r);
        if (!i.isBlockContainer || !i.childContainer)
          return !1;
        const { blockContent: l, childContainer: a } = i, c = r.selection.from === l.afterPos - 1, d = r.selection.empty, u = v(
          r.doc.resolve(a.beforePos + 1)
        );
        if (!u.isBlockContainer)
          return !1;
        if (c && d) {
          const h = u.blockContent.node, f = h.type.spec.content === "inline*", m = l.node.type.spec.content === "inline*";
          return n().insertContentAt(
            u.bnBlock.afterPos,
            ((p = u.childContainer) == null ? void 0 : p.node.content) || M.empty
          ).deleteRange(
            // Deletes whole child container if there's only one child.
            a.node.childCount === 1 ? {
              from: a.beforePos,
              to: a.afterPos
            } : {
              from: u.bnBlock.beforePos,
              to: u.bnBlock.afterPos
            }
          ).insertContentAt(
            r.selection.from,
            f && m ? h.content : null
          ).setTextSelection(r.selection.from).scrollIntoView().run();
        }
        return !1;
      }),
      // Merges block with the next one (at the same nesting level or lower),
      // if one exists, the block has no children, and the selection is at the
      // end of the block.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const { bnBlock: l, blockContent: a } = i, c = L(
          r.doc,
          i.bnBlock.beforePos
        );
        if (!c || !c.isBlockContainer)
          return !1;
        const d = r.selection.from === a.afterPos - 1, u = r.selection.empty, p = l.afterPos;
        return d && u ? n().command(fe(p)).scrollIntoView().run() : !1;
      }),
      // If the previous block is a columnList, moves the current block to
      // the end of the last column in it.
      () => s.command(({ state: r, tr: i, dispatch: l }) => {
        const a = B(r);
        if (!a.isBlockContainer)
          return !1;
        const c = L(
          r.doc,
          a.bnBlock.beforePos
        );
        if (!c || c.isBlockContainer)
          return !1;
        if (l) {
          const d = c.bnBlock.beforePos + 1, u = i.doc.resolve(d + 1);
          return i.delete(
            u.pos,
            u.pos + u.nodeAfter.nodeSize
          ), H(i, c.bnBlock.beforePos), i.insert(a.bnBlock.afterPos, u.nodeAfter), i.setSelection(
            S.near(i.doc.resolve(u.pos))
          ), !0;
        }
        return !1;
      }),
      // If the block is the last in a column, moves it to the start of the
      // next column. If there is no next column, moves it below the
      // columnList.
      () => s.command(({ state: r, tr: i, dispatch: l }) => {
        const a = B(r);
        if (!a.isBlockContainer || !(i.selection.from === a.blockContent.afterPos - 1))
          return !1;
        const d = i.doc.resolve(a.bnBlock.afterPos);
        if (d.nodeAfter || d.node().type.name !== "column")
          return !1;
        const h = i.doc.resolve(a.bnBlock.afterPos), f = i.doc.resolve(h.after()), m = f.after();
        if (l) {
          const b = f.pos === m - 1 ? m : f.pos + 1, g = v(
            i.doc.resolve(b)
          );
          i.delete(
            g.bnBlock.beforePos,
            g.bnBlock.afterPos
          ), H(
            i,
            m - f.node().nodeSize
          ), i.insert(h.pos, g.bnBlock.node), i.setSelection(
            S.near(i.doc.resolve(b))
          );
        }
        return !0;
      }),
      // Deletes the next block at either the same or lower nesting level, if
      // the selection is empty and at the end of the block. If both the
      // current and next blocks have inline content, the next block's
      // content is appended to the current block's. The next block's own
      // children are unindented before it's deleted.
      () => s.command(({ state: r }) => {
        var d;
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const { blockContent: l } = i, a = r.selection.from === l.afterPos - 1, c = r.selection.empty;
        if (a && c) {
          const u = (b, g) => {
            const x = L(b, g);
            if (x)
              return x;
            const y = Le(b, g);
            if (y)
              return u(
                b,
                y.bnBlock.beforePos
              );
          }, p = u(
            r.doc,
            i.bnBlock.beforePos
          );
          if (!p || !p.isBlockContainer)
            return !1;
          const h = p.blockContent.node, f = h.type.spec.content === "inline*", m = l.node.type.spec.content === "inline*";
          return n().insertContentAt(
            p.bnBlock.afterPos,
            ((d = p.childContainer) == null ? void 0 : d.node.content) || M.empty
          ).deleteRange({
            from: p.bnBlock.beforePos,
            to: p.bnBlock.afterPos
          }).insertContentAt(
            r.selection.from,
            f && m ? h.content : null
          ).setTextSelection(r.selection.from).scrollIntoView().run();
        }
        return !1;
      }),
      // Deletes the current block if it's an empty block with inline content,
      // and moves the selection to the next block.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        if (i.blockContent.node.childCount === 0 && i.blockContent.node.type.spec.content === "inline*") {
          const a = L(
            r.doc,
            i.bnBlock.beforePos
          );
          if (!a || !a.isBlockContainer)
            return !1;
          let c = n();
          if (a.blockContent.node.type.spec.content === "tableRow+") {
            const f = i.bnBlock.afterPos + 1 + 1 + 1 + 1 + 1;
            c = c.setTextSelection(
              f
            );
          } else a.blockContent.node.type.spec.content === "" ? c = c.setNodeSelection(
            a.blockContent.beforePos
          ) : c = c.setTextSelection(
            a.blockContent.beforePos + 1
          );
          return c.deleteRange({
            from: i.bnBlock.beforePos,
            to: i.bnBlock.afterPos
          }).scrollIntoView().run();
        }
        return !1;
      }),
      // Deletes next block if it contains no content and isn't a table,
      // when the selection is empty and at the end of the block. Moves the
      // current block into the deleted block's place.
      () => s.command(({ state: r }) => {
        const i = B(r);
        if (!i.isBlockContainer)
          return !1;
        const l = r.selection.from === i.blockContent.afterPos - 1, a = r.selection.empty, c = L(
          r.doc,
          i.bnBlock.beforePos
        );
        if (!c || !c.isBlockContainer)
          return !1;
        if (c && l && a && (c.blockContent.node.type.spec.content === "" || c.blockContent.node.type.spec.content === "inline*" && c.blockContent.node.childCount === 0)) {
          const u = c.bnBlock.node.lastChild.content;
          return n().deleteRange({
            from: c.bnBlock.beforePos,
            to: c.bnBlock.afterPos
          }).insertContentAt(
            i.bnBlock.afterPos,
            c.bnBlock.node.childCount === 2 ? u : null
          ).run();
        }
        return !1;
      })
    ]), t = (n = !1) => this.editor.commands.first(({ commands: s, tr: r }) => [
      // Removes a level of nesting if the block is empty & indented, while the selection is also empty & at the start
      // of the block.
      () => s.command(({ state: i }) => {
        const l = B(i);
        if (!l.isBlockContainer)
          return !1;
        const { bnBlock: a, blockContent: c } = l, { depth: d } = i.doc.resolve(a.beforePos), u = i.selection.$anchor.parentOffset === 0, p = i.selection.anchor === i.selection.head, h = c.node.childCount === 0, f = d > 1;
        return u && p && h && f ? s.liftListItem("blockContainer") : !1;
      }),
      // Creates a hard break if block is configured to do so.
      () => s.command(({ state: i }) => {
        var c;
        const l = B(i), a = ((c = this.options.editor.schema.blockSchema[l.blockNoteType].meta) == null ? void 0 : c.hardBreakShortcut) ?? "shift+enter";
        if (a === "none")
          return !1;
        if (
          // If shortcut is not configured, or is configured as "shift+enter",
          // create a hard break for shift+enter, but not for enter.
          a === "shift+enter" && n || // If shortcut is configured as "enter", create a hard break for
          // both enter and shift+enter.
          a === "enter"
        ) {
          const d = r.storedMarks || r.selection.$head.marks().filter(
            (u) => this.editor.extensionManager.splittableMarks.includes(
              u.type.name
            )
          );
          return r.insert(
            r.selection.head,
            r.doc.type.schema.nodes.hardBreak.create()
          ).ensureMarks(d), !0;
        }
        return !1;
      }),
      // Creates a new block and moves the selection to it if the current one is empty, while the selection is also
      // empty & at the start of the block.
      () => s.command(({ state: i, dispatch: l, tr: a }) => {
        var m;
        const c = B(i);
        if (!c.isBlockContainer)
          return !1;
        const { bnBlock: d, blockContent: u } = c, p = i.selection.$anchor.parentOffset === 0, h = i.selection.anchor === i.selection.head, f = u.node.childCount === 0;
        if (p && h && f) {
          const b = d.afterPos, g = b + 2;
          if (l) {
            const x = i.schema.nodes.blockContainer.createAndFill(
              void 0,
              [
                i.schema.nodes.paragraph.createAndFill() || void 0,
                (m = c.childContainer) == null ? void 0 : m.node
              ].filter((y) => y !== void 0)
            );
            a.insert(b, x).setSelection(
              new S(a.doc.resolve(g))
            ).scrollIntoView(), c.childContainer && a.delete(
              c.childContainer.beforePos,
              c.childContainer.afterPos
            );
          }
          return !0;
        }
        return !1;
      }),
      // Splits the current block, moving content inside that's after the cursor to a new text block below. Also
      // deletes the selection beforehand, if it's not empty.
      () => s.command(({ state: i, chain: l }) => {
        const a = B(i);
        if (!a.isBlockContainer)
          return !1;
        const { blockContent: c } = a, d = i.selection.$anchor.parentOffset === 0;
        return c.node.childCount === 0 ? !1 : (l().deleteSelection().command(
          ot(
            i.selection.from,
            d,
            d
          )
        ).run(), !0);
      })
    ]);
    return {
      Backspace: o,
      Delete: e,
      Enter: () => t(),
      "Shift-Enter": () => t(!0),
      // Always returning true for tab key presses ensures they're not captured by the browser. Otherwise, they blur the
      // editor since the browser will try to use tab for keyboard navigation.
      Tab: () => {
        var n, s;
        return this.options.tabBehavior !== "prefer-indent" && ((n = this.options.editor.getExtension(Y)) != null && n.store.state || ((s = this.options.editor.getExtension(J)) == null ? void 0 : s.store.state) !== void 0) ? !1 : Me(this.options.editor);
      },
      "Shift-Tab": () => {
        var n, s;
        return this.options.tabBehavior !== "prefer-indent" && ((n = this.options.editor.getExtension(Y)) != null && n.store.state || ((s = this.options.editor.getExtension(J)) == null ? void 0 : s.store.state) !== void 0) ? !1 : this.editor.commands.liftListItem("blockContainer");
      },
      "Shift-Mod-ArrowUp": () => (this.options.editor.moveBlocksUp(), !0),
      "Shift-Mod-ArrowDown": () => (this.options.editor.moveBlocksDown(), !0),
      "Mod-z": () => this.options.editor.undo(),
      "Mod-y": () => this.options.editor.redo(),
      "Shift-Mod-z": () => this.options.editor.redo()
    };
  }
}), pn = te.create({
  name: "insertion",
  inclusive: !1,
  excludes: "deletion modification insertion",
  addAttributes() {
    return {
      id: { default: null, validate: "number" }
      // note: validate is supported in prosemirror but not in tiptap, so this doesn't actually work (considered not critical)
    };
  },
  extendMarkSchema(o) {
    return o.name !== "insertion" ? {} : {
      blocknoteIgnore: !0,
      inclusive: !1,
      toDOM(e, t) {
        return [
          "ins",
          {
            "data-id": String(e.attrs.id),
            "data-inline": String(t),
            ...!t && { style: "display: contents" }
            // changed to "contents" to make this work for table rows
          },
          0
        ];
      },
      parseDOM: [
        {
          tag: "ins",
          getAttrs(e) {
            return e.dataset.id ? {
              id: parseInt(e.dataset.id, 10)
            } : !1;
          }
        }
      ]
    };
  }
}), fn = te.create({
  name: "deletion",
  inclusive: !1,
  excludes: "insertion modification deletion",
  addAttributes() {
    return {
      id: { default: null, validate: "number" }
      // note: validate is supported in prosemirror but not in tiptap
    };
  },
  extendMarkSchema(o) {
    return o.name !== "deletion" ? {} : {
      blocknoteIgnore: !0,
      inclusive: !1,
      // attrs: {
      //   id: { validate: "number" },
      // },
      toDOM(e, t) {
        return [
          "del",
          {
            "data-id": String(e.attrs.id),
            "data-inline": String(t),
            ...!t && { style: "display: contents" }
            // changed to "contents" to make this work for table rows
          },
          0
        ];
      },
      parseDOM: [
        {
          tag: "del",
          getAttrs(e) {
            return e.dataset.id ? {
              id: parseInt(e.dataset.id, 10)
            } : !1;
          }
        }
      ]
    };
  }
}), hn = te.create({
  name: "modification",
  inclusive: !1,
  excludes: "deletion insertion",
  addAttributes() {
    return {
      id: { default: null, validate: "number" },
      type: { validate: "string" },
      attrName: { default: null, validate: "string|null" },
      previousValue: { default: null },
      newValue: { default: null }
    };
  },
  extendMarkSchema(o) {
    return o.name !== "modification" ? {} : {
      blocknoteIgnore: !0,
      inclusive: !1,
      // attrs: {
      //   id: { validate: "number" },
      //   type: { validate: "string" },
      //   attrName: { default: null, validate: "string|null" },
      //   previousValue: { default: null },
      //   newValue: { default: null },
      // },
      toDOM(e, t) {
        return [
          t ? "span" : "div",
          {
            "data-type": "modification",
            "data-id": String(e.attrs.id),
            "data-mod-type": e.attrs.type,
            "data-mod-prev-val": JSON.stringify(e.attrs.previousValue),
            // TODO: Try to serialize marks with toJSON?
            "data-mod-new-val": JSON.stringify(e.attrs.newValue)
          },
          0
        ];
      },
      parseDOM: [
        {
          tag: "span[data-type='modification']",
          getAttrs(e) {
            return e.dataset.id ? {
              id: parseInt(e.dataset.id, 10),
              type: e.dataset.modType,
              previousValue: e.dataset.modPrevVal,
              newValue: e.dataset.modNewVal
            } : !1;
          }
        },
        {
          tag: "div[data-type='modification']",
          getAttrs(e) {
            return e.dataset.id ? {
              id: parseInt(e.dataset.id, 10),
              type: e.dataset.modType,
              previousValue: e.dataset.modPrevVal
            } : !1;
          }
        }
      ]
    };
  }
}), mn = T.create({
  name: "textAlignment",
  addGlobalAttributes() {
    return [
      {
        // Generally text alignment is handled through props using the custom
        // blocks API. Tables are the only blocks that are created as TipTap
        // nodes and ported to blocks, so we need to add text alignment in a
        // separate extension.
        types: ["tableCell", "tableHeader"],
        attributes: {
          textAlignment: {
            default: "left",
            parseHTML: (o) => o.getAttribute("data-text-alignment"),
            renderHTML: (o) => o.textAlignment === "left" ? {} : {
              "data-text-alignment": o.textAlignment
            }
          }
        }
      }
    ];
  }
}), kn = T.create({
  name: "blockTextColor",
  addGlobalAttributes() {
    return [
      {
        types: ["table", "tableCell", "tableHeader"],
        attributes: {
          textColor: st()
        }
      }
    ];
  }
}), bn = {
  blockColor: "data-block-color",
  blockStyle: "data-block-style",
  id: "data-id",
  depth: "data-depth",
  depthChange: "data-depth-change"
}, gn = O.create({
  name: "blockContainer",
  group: "blockGroupChild bnBlock",
  // A block always contains content, and optionally a blockGroup which contains nested blocks
  content: "blockContent blockGroup?",
  // Ensures content-specific keyboard handlers trigger first.
  priority: 50,
  defining: !0,
  marks: "insertion modification deletion",
  parseHTML() {
    return [
      {
        tag: "div[data-node-type=" + this.name + "]",
        getAttrs: (o) => {
          if (typeof o == "string")
            return !1;
          const e = {};
          for (const [t, n] of Object.entries(bn))
            o.getAttribute(n) && (e[t] = o.getAttribute(n));
          return e;
        }
      },
      // Ignore `blockOuter` divs, but parse the `blockContainer` divs inside them.
      {
        tag: 'div[data-node-type="blockOuter"]',
        skip: !0
      }
    ];
  },
  renderHTML({ HTMLAttributes: o }) {
    var s;
    const e = document.createElement("div");
    e.className = "bn-block-outer", e.setAttribute("data-node-type", "blockOuter");
    for (const [r, i] of Object.entries(o))
      r !== "class" && e.setAttribute(r, i);
    const t = {
      ...((s = this.options.domAttributes) == null ? void 0 : s.block) || {},
      ...o
    }, n = document.createElement("div");
    n.className = ee("bn-block", t.class), n.setAttribute("data-node-type", this.name);
    for (const [r, i] of Object.entries(t))
      r !== "class" && n.setAttribute(r, i);
    return e.appendChild(n), {
      dom: e,
      contentDOM: n
    };
  }
}), Bn = O.create({
  name: "blockGroup",
  group: "childContainer",
  content: "blockGroupChild+",
  marks: "deletion insertion modification",
  parseHTML() {
    return [
      {
        tag: "div",
        getAttrs: (o) => typeof o == "string" ? !1 : o.getAttribute("data-node-type") === "blockGroup" ? null : !1
      }
    ];
  },
  renderHTML({ HTMLAttributes: o }) {
    var n;
    const e = {
      ...((n = this.options.domAttributes) == null ? void 0 : n.blockGroup) || {},
      ...o
    }, t = document.createElement("div");
    t.className = ee(
      "bn-block-group",
      e.class
    ), t.setAttribute("data-node-type", "blockGroup");
    for (const [s, r] of Object.entries(e))
      s !== "class" && t.setAttribute(s, r);
    return {
      dom: t,
      contentDOM: t
    };
  }
}), yn = O.create({
  name: "doc",
  topNode: !0,
  content: "blockGroup",
  marks: "insertion modification deletion"
}), Cn = Vt(
  ({ options: o }) => ({
    key: "collaboration",
    blockNoteExtensions: [
      dt(o),
      ut(o),
      pt(o),
      ft(),
      ht(o)
    ]
  })
);
let he = !1;
function Sn(o, e) {
  const t = [
    N.ClipboardTextSerializer,
    N.Commands,
    N.Editable,
    N.FocusEvents,
    N.Tabindex,
    zt,
    ke.configure({
      // everything from bnBlock group (nodes that represent a BlockNote block should have an id)
      types: ["blockContainer", "columnList", "column"],
      setIdAttribute: e.setIdAttribute
    }),
    cn,
    Gt,
    // marks:
    pn,
    fn,
    hn,
    Rt.extend({
      inclusive: !1
    }).configure({
      defaultProtocol: Pt,
      // only call this once if we have multiple editors installed. Or fix https://github.com/ueberdosis/tiptap/issues/5450
      protocols: he ? [] : Et
    }),
    ...Object.values(o.schema.styleSpecs).map((n) => n.implementation.mark.configure({
      editor: o
    })),
    kn,
    an,
    mn,
    // make sure escape blurs editor, so that we can tab to other elements in the host page (accessibility)
    T.create({
      name: "OverrideEscape",
      addKeyboardShortcuts: () => ({
        Escape: () => {
          var n;
          return (n = o.getExtension(ge)) != null && n.shown() ? !1 : (o.blur(), !0);
        }
      })
    }),
    // nodes
    yn,
    gn.configure({
      editor: o,
      domAttributes: e.domAttributes
    }),
    un.configure({
      editor: o,
      tabBehavior: e.tabBehavior
    }),
    Bn.configure({
      domAttributes: e.domAttributes
    }),
    ...Object.values(o.schema.inlineContentSpecs).filter((n) => n.config !== "link" && n.config !== "text").map((n) => n.implementation.node.configure({
      editor: o
    })),
    ...Object.values(o.schema.blockSpecs).flatMap((n) => [
      // the node extension implementations
      ..."node" in n.implementation ? [
        n.implementation.node.configure({
          editor: o,
          domAttributes: e.domAttributes
        })
      ] : []
    ]),
    rn(o),
    nn(
      o,
      e.pasteHandler || ((n) => n.defaultPasteHandler())
    ),
    Vo(o)
  ];
  return he = !0, t;
}
function xn(o, e) {
  const t = [
    ye(),
    mt(e),
    J(e),
    Y(e),
    kt(e),
    bt(),
    gt(e),
    Wt(e),
    Bt(e),
    ge(e),
    ...e.trailingBlock !== !1 ? [yt()] : []
  ];
  return e.collaboration ? t.push(Cn(e.collaboration)) : t.push(Ct()), "table" in o.schema.blockSpecs && t.push(St(e)), e.animations !== !1 && t.push(xt()), t;
}
class En {
  constructor(e, t) {
    /**
     * A set of extension keys which are disabled by the options
     */
    k(this, "disabledExtensions", /* @__PURE__ */ new Set());
    /**
     * A list of all the extensions that are registered to the editor
     */
    k(this, "extensions", []);
    /**
     * A map of all the abort controllers for each extension that has an init method defined
     */
    k(this, "abortMap", /* @__PURE__ */ new Map());
    /**
     * A map of all the extension factories that are registered to the editor
     */
    k(this, "extensionFactories", /* @__PURE__ */ new Map());
    /**
     * Because a single blocknote extension can both have it's own prosemirror plugins & additional generated ones (e.g. keymap & input rules plugins)
     * We need to keep track of all the plugins for each extension, so that we can remove them when the extension is unregistered
     */
    k(this, "extensionPlugins", /* @__PURE__ */ new Map());
    this.editor = e, this.options = t, e.onMount(() => {
      for (const n of this.extensions)
        if (n.mount) {
          const s = new window.AbortController(), r = n.mount({
            dom: e.prosemirrorView.dom,
            root: e.prosemirrorView.root,
            signal: s.signal
          });
          r && s.signal.addEventListener("abort", () => {
            r();
          }), this.abortMap.set(n, s);
        }
    }), e.onUnmount(() => {
      for (const [n, s] of this.abortMap.entries())
        this.abortMap.delete(n), s.abort();
    }), this.disabledExtensions = new Set(t.disableExtensions || []);
    for (const n of xn(this.editor, this.options))
      this.addExtension(n);
    for (const n of this.options.extensions ?? [])
      this.addExtension(n);
    for (const n of Object.values(this.editor.schema.blockSpecs))
      for (const s of n.extensions ?? [])
        this.addExtension(s);
  }
  /**
   * Register one or more extensions to the editor after the editor is initialized.
   *
   * This allows users to switch on & off extensions "at runtime".
   */
  registerExtension(e) {
    var r;
    const t = [].concat(e).filter(Boolean);
    if (!t.length) {
      console.warn("No extensions found to register", e);
      return;
    }
    const n = t.map((i) => this.addExtension(i)).filter(Boolean), s = /* @__PURE__ */ new Set();
    for (const i of n)
      i != null && i.tiptapExtensions && console.warn(
        `Extension ${i.key} has tiptap extensions, but these cannot be changed after initializing the editor. Please separate the extension into multiple extensions if you want to add them, or re-initialize the editor.`,
        i
      ), (r = i == null ? void 0 : i.inputRules) != null && r.length && console.warn(
        `Extension ${i.key} has input rules, but these cannot be changed after initializing the editor. Please separate the extension into multiple extensions if you want to add them, or re-initialize the editor.`,
        i
      ), this.getProsemirrorPluginsFromExtension(i).plugins.forEach(
        (l) => {
          s.add(l);
        }
      );
    this.updatePlugins((i) => [...i, ...s]);
  }
  /**
   * Register an extension to the editor
   * @param extension - The extension to register
   * @returns The extension instance
   */
  addExtension(e) {
    let t;
    if (typeof e == "function" ? t = e({ editor: this.editor }) : t = e, !(!t || this.disabledExtensions.has(t.key))) {
      if (typeof e == "function") {
        const n = t[Ut];
        typeof n == "function" && this.extensionFactories.set(n, t);
      }
      if (this.extensions.push(t), t.blockNoteExtensions)
        for (const n of t.blockNoteExtensions)
          this.addExtension(n);
      return t;
    }
  }
  /**
   * Resolve an extension or a list of extensions into a list of extension instances
   * @param toResolve - The extension or list of extensions to resolve
   * @returns A list of extension instances
   */
  resolveExtensions(e) {
    const t = [];
    if (typeof e == "function") {
      const n = this.extensionFactories.get(e);
      n && t.push(n);
    } else if (Array.isArray(e))
      for (const n of e)
        t.push(...this.resolveExtensions(n));
    else if (typeof e == "object" && "key" in e)
      t.push(e);
    else if (typeof e == "string") {
      const n = this.extensions.find((s) => s.key === e);
      n && t.push(n);
    }
    return t;
  }
  /**
   * Unregister an extension from the editor
   * @param toUnregister - The extension to unregister
   * @returns void
   */
  unregisterExtension(e) {
    var r;
    const t = this.resolveExtensions(e);
    if (!t.length) {
      console.warn("No extensions found to unregister", e);
      return;
    }
    let n = !1;
    const s = /* @__PURE__ */ new Set();
    for (const i of t) {
      this.extensions = this.extensions.filter((a) => a !== i), this.extensionFactories.forEach((a, c) => {
        a === i && this.extensionFactories.delete(c);
      }), (r = this.abortMap.get(i)) == null || r.abort(), this.abortMap.delete(i);
      const l = this.extensionPlugins.get(i);
      l == null || l.forEach((a) => {
        s.add(a);
      }), this.extensionPlugins.delete(i), i.tiptapExtensions && !n && (n = !0, console.warn(
        `Extension ${i.key} has tiptap extensions, but they will not be removed. Please separate the extension into multiple extensions if you want to remove them, or re-initialize the editor.`,
        e
      ));
    }
    this.updatePlugins(
      (i) => i.filter((l) => !s.has(l))
    );
  }
  /**
   * Allows resetting the current prosemirror state's plugins
   * @param update - A function that takes the current plugins and returns the new plugins
   * @returns void
   */
  updatePlugins(e) {
    const t = this.editor.prosemirrorState, n = t.reconfigure({
      plugins: e(t.plugins.slice())
    });
    this.editor.prosemirrorView.updateState(n);
  }
  /**
   * Get all the extensions that are registered to the editor
   */
  getTiptapExtensions() {
    var s;
    const e = Sn(
      this.editor,
      this.options
    ).filter((r) => !this.disabledExtensions.has(r.name)), t = Tt(this.extensions), n = /* @__PURE__ */ new Map();
    for (const r of this.extensions) {
      r.tiptapExtensions && e.push(...r.tiptapExtensions);
      const i = t(r.key), { plugins: l, inputRules: a } = this.getProsemirrorPluginsFromExtension(r);
      l.length && e.push(
        T.create({
          name: r.key,
          priority: i,
          addProseMirrorPlugins: () => l
        })
      ), a.length && (n.has(i) || n.set(i, []), n.get(i).push(...a));
    }
    e.push(
      T.create({
        name: "blocknote-input-rules",
        addProseMirrorPlugins() {
          const r = [];
          return Array.from(n.keys()).sort().reverse().forEach((i) => {
            r.push(...n.get(i));
          }), [$t({ rules: r })];
        }
      })
    );
    for (const r of ((s = this.options._tiptapOptions) == null ? void 0 : s.extensions) ?? [])
      e.push(r);
    return e;
  }
  /**
   * This maps a blocknote extension into an array of Prosemirror plugins if it has any of the following:
   * - plugins
   * - keyboard shortcuts
   * - input rules
   */
  getProsemirrorPluginsFromExtension(e) {
    var s, r, i;
    const t = [...e.prosemirrorPlugins ?? []], n = [];
    return !((s = e.prosemirrorPlugins) != null && s.length) && !Object.keys(e.keyboardShortcuts || {}).length && !((r = e.inputRules) != null && r.length) ? { plugins: t, inputRules: n } : (this.extensionPlugins.set(e, t), (i = e.inputRules) != null && i.length && n.push(
      ...e.inputRules.map((l) => new Ht(l.find, (a, c, d, u) => {
        const p = l.replace({
          match: c,
          range: { from: d, to: u },
          editor: this.editor
        });
        if (p) {
          const h = this.editor.getTextCursorPosition();
          if (this.editor.schema.blockSchema[h.block.type].content !== "inline")
            return null;
          const f = z(a.tr), m = a.tr.deleteRange(d, u);
          return rt(m, f.bnBlock.beforePos, p), m;
        }
        return null;
      }))
    ), Object.keys(e.keyboardShortcuts || {}).length && t.push(
      Ft(
        Object.fromEntries(
          Object.entries(e.keyboardShortcuts).map(([l, a]) => [
            l,
            () => a({ editor: this.editor })
          ])
        )
      )
    ), { plugins: t, inputRules: n });
  }
  /**
   * Get all extensions
   */
  getExtensions() {
    return new Map(
      this.extensions.map((e) => [e.key, e])
    );
  }
  getExtension(e) {
    if (typeof e == "string") {
      const t = this.extensions.find((n) => n.key === e);
      return t || void 0;
    } else if (typeof e == "function") {
      const t = this.extensionFactories.get(e);
      return t || void 0;
    }
    throw new Error(`Invalid extension type: ${typeof e}`);
  }
  /**
   * Check if an extension exists
   */
  hasExtension(e) {
    return typeof e == "string" ? this.extensions.some((t) => t.key === e) : typeof e == "object" && "key" in e ? this.extensions.some((t) => t.key === e.key) : typeof e == "function" ? this.extensionFactories.has(e) : !1;
  }
}
function Pn(o, e) {
  let { $from: t, $to: n } = e;
  if (t.pos > t.start() && t.pos < o.content.size) {
    const s = o.textBetween(t.pos, t.pos + 1);
    if (/^[\w\p{P}]$/u.test(s)) {
      const i = o.textBetween(t.start(), t.pos).match(/[\w\p{P}]+$/u);
      i && (t = o.resolve(t.pos - i[0].length));
    }
  }
  if (n.pos < n.end() && n.pos > 0) {
    const s = o.textBetween(n.pos - 1, n.pos);
    if (/^[\w\p{P}]$/u.test(s)) {
      const i = o.textBetween(n.pos, n.end()).match(/^[\w\p{P}]+/u);
      i && (n = o.resolve(n.pos + i[0].length));
    }
  }
  return { $from: t, $to: n, from: t.pos, to: n.pos };
}
function Tn(o) {
  const e = E(o);
  if (o.selection.empty || "node" in o.selection)
    return;
  const t = o.doc.resolve(
    D(o.doc, o.selection.from).posBeforeNode
  ), n = o.doc.resolve(
    D(o.doc, o.selection.to).posBeforeNode
  ), s = (c, d) => {
    const u = t.posAtIndex(c, d), p = o.doc.resolve(u).nodeAfter;
    if (!p)
      throw new Error(
        `Error getting selection - node not found at position ${u}`
      );
    return C(p, e);
  }, r = [], i = t.sharedDepth(n.pos), l = t.index(i), a = n.index(i);
  if (t.depth > i) {
    r.push(C(t.nodeAfter, e));
    for (let c = t.depth; c > i; c--)
      if (t.node(c).type.isInGroup("childContainer")) {
        const u = t.index(c) + 1, p = t.node(c).childCount;
        for (let h = u; h < p; h++)
          r.push(s(h, c));
      }
  } else
    r.push(s(l, i));
  for (let c = l + 1; c <= a; c++)
    r.push(s(c, i));
  if (r.length === 0)
    throw new Error(
      `Error getting selection - selection doesn't span any blocks (${o.selection})`
    );
  return {
    blocks: r
  };
}
function wn(o, e, t) {
  const n = typeof e == "string" ? e : e.id, s = typeof t == "string" ? t : t.id, r = E(o), i = be(r);
  if (n === s)
    throw new Error(
      `Attempting to set selection with the same anchor and head blocks (id ${n})`
    );
  const l = P(n, o.doc);
  if (!l)
    throw new Error(`Block with ID ${n} not found`);
  const a = P(s, o.doc);
  if (!a)
    throw new Error(`Block with ID ${s} not found`);
  const c = K(l), d = K(a), u = i.blockSchema[c.blockNoteType], p = i.blockSchema[d.blockNoteType];
  if (!c.isBlockContainer || u.content === "none")
    throw new Error(
      `Attempting to set selection anchor in block without content (id ${n})`
    );
  if (!d.isBlockContainer || p.content === "none")
    throw new Error(
      `Attempting to set selection anchor in block without content (id ${s})`
    );
  let h, f;
  if (u.content === "table") {
    const m = re.get(c.blockContent.node);
    h = c.blockContent.beforePos + m.positionAt(0, 0, c.blockContent.node) + 1 + 2;
  } else
    h = c.blockContent.beforePos + 1;
  if (p.content === "table") {
    const m = re.get(d.blockContent.node), b = d.blockContent.beforePos + m.positionAt(
      m.height - 1,
      m.width - 1,
      d.blockContent.node
    ) + 1, g = o.doc.resolve(b).nodeAfter.nodeSize;
    f = b + g - 2;
  } else
    f = d.blockContent.afterPos - 1;
  o.setSelection(S.create(o.doc, h, f));
}
function Mn(o, e = !1) {
  const t = E(o), n = e ? Pn(o.doc, o.selection) : o.selection;
  let s = n.$from, r = n.$to;
  for (; r.parentOffset >= r.parent.nodeSize - 2 && r.depth > 0; )
    r = o.doc.resolve(r.pos + 1);
  for (; r.parentOffset === 0 && r.depth > 0; )
    r = o.doc.resolve(r.pos - 1);
  for (; s.parentOffset === 0 && s.depth > 0; )
    s = o.doc.resolve(s.pos - 1);
  for (; s.parentOffset >= s.parent.nodeSize - 2 && s.depth > 0; )
    s = o.doc.resolve(s.pos + 1);
  const i = Ke(
    o.doc.slice(s.pos, r.pos, !0),
    t
  );
  return {
    _meta: {
      startPos: s.pos,
      endPos: r.pos
    },
    ...i
  };
}
function vn(o) {
  const { bnBlock: e } = z(o), t = E(o.doc), n = o.doc.resolve(e.beforePos), s = n.nodeBefore, r = o.doc.resolve(e.afterPos).nodeAfter;
  let i;
  return n.depth > 1 && (i = n.node(), i.type.isInGroup("bnBlock") || (i = n.node(n.depth - 1))), {
    block: C(e.node, t),
    prevBlock: s === null ? void 0 : C(s, t),
    nextBlock: r === null ? void 0 : C(r, t),
    parentBlock: i === void 0 ? void 0 : C(i, t)
  };
}
function De(o, e, t = "start") {
  const n = typeof e == "string" ? e : e.id, s = E(o.doc), r = be(s), i = P(n, o.doc);
  if (!i)
    throw new Error(`Block with ID ${n} not found`);
  const l = K(i), a = r.blockSchema[l.blockNoteType].content;
  if (l.isBlockContainer) {
    const c = l.blockContent;
    if (a === "none") {
      o.setSelection($.create(o.doc, c.beforePos));
      return;
    }
    if (a === "inline")
      t === "start" ? o.setSelection(
        S.create(o.doc, c.beforePos + 1)
      ) : o.setSelection(
        S.create(o.doc, c.afterPos - 1)
      );
    else if (a === "table")
      t === "start" ? o.setSelection(
        S.create(o.doc, c.beforePos + 4)
      ) : o.setSelection(
        S.create(o.doc, c.afterPos - 4)
      );
    else
      throw new U(a);
  } else {
    const c = t === "start" ? l.childContainer.node.firstChild : l.childContainer.node.lastChild;
    De(o, c.attrs.id, t);
  }
}
class In {
  constructor(e) {
    this.editor = e;
  }
  /**
   * Gets a snapshot of the current selection. This contains all blocks (included nested blocks)
   * that the selection spans across.
   *
   * If the selection starts / ends halfway through a block, the returned data will contain the entire block.
   */
  getSelection() {
    return this.editor.transact((e) => Tn(e));
  }
  /**
   * Gets a snapshot of the current selection. This contains all blocks (included nested blocks)
   * that the selection spans across.
   *
   * If the selection starts / ends halfway through a block, the returned block will be
   * only the part of the block that is included in the selection.
   */
  getSelectionCutBlocks(e = !1) {
    return this.editor.transact(
      (t) => Mn(t, e)
    );
  }
  /**
   * Sets the selection to a range of blocks.
   * @param startBlock The identifier of the block that should be the start of the selection.
   * @param endBlock The identifier of the block that should be the end of the selection.
   */
  setSelection(e, t) {
    return this.editor.transact((n) => wn(n, e, t));
  }
  /**
   * Gets a snapshot of the current text cursor position.
   * @returns A snapshot of the current text cursor position.
   */
  getTextCursorPosition() {
    return this.editor.transact((e) => vn(e));
  }
  /**
   * Sets the text cursor position to the start or end of an existing block. Throws an error if the target block could
   * not be found.
   * @param targetBlock The identifier of an existing block that the text cursor should be moved to.
   * @param placement Whether the text cursor should be placed at the start or end of the block.
   */
  setTextCursorPosition(e, t = "start") {
    return this.editor.transact(
      (n) => De(n, e, t)
    );
  }
  /**
   * Gets the bounding box of the current selection.
   */
  getSelectionBoundingBox() {
    if (!this.editor.prosemirrorView)
      return;
    const { selection: e } = this.editor.prosemirrorState, { ranges: t } = e, n = Math.min(...t.map((r) => r.$from.pos)), s = Math.max(...t.map((r) => r.$to.pos));
    if (vt(e)) {
      const r = this.editor.prosemirrorView.nodeDOM(n);
      if (r)
        return r.getBoundingClientRect();
    }
    return It(
      this.editor.prosemirrorView,
      n,
      s
    ).toJSON();
  }
}
class An {
  constructor(e) {
    /**
     * Stores the currently active transaction, which is the accumulated transaction from all {@link dispatch} calls during a {@link transact} calls
     */
    k(this, "activeTransaction", null);
    // Flag to indicate if we're in a `can` call
    k(this, "isInCan", !1);
    this.editor = e;
  }
  /**
   * For any command that can be executed, you can check if it can be executed by calling `editor.can(command)`.
   * @example
   * ```ts
   * if (editor.can(editor.undo)) {
   *   // show button
   * } else {
   *   // hide button
   * }
   */
  can(e) {
    try {
      return this.isInCan = !0, e();
    } finally {
      this.isInCan = !1;
    }
  }
  /**
   * Execute a prosemirror command. This is mostly for backwards compatibility with older code.
   *
   * @note You should prefer the {@link transact} method when possible, as it will automatically handle the dispatching of the transaction and work across blocknote transactions.
   *
   * @example
   * ```ts
   * editor.exec((state, dispatch, view) => {
   *   dispatch(state.tr.insertText("Hello, world!"));
   * });
   * ```
   */
  exec(e) {
    if (this.activeTransaction)
      throw new Error(
        "`exec` should not be called within a `transact` call, move the `exec` call outside of the `transact` call"
      );
    if (this.isInCan)
      return this.canExec(e);
    const t = this.prosemirrorState, n = this.prosemirrorView;
    return e(t, (r) => this.prosemirrorView.dispatch(r), n);
  }
  /**
   * Check if a command can be executed. A command should return `false` if it is not valid in the current state.
   *
   * @example
   * ```ts
   * if (editor.canExec(command)) {
   *   // show button
   * } else {
   *   // hide button
   * }
   * ```
   */
  canExec(e) {
    if (this.activeTransaction)
      throw new Error(
        "`canExec` should not be called within a `transact` call, move the `canExec` call outside of the `transact` call"
      );
    const t = this.prosemirrorState, n = this.prosemirrorView;
    return e(t, void 0, n);
  }
  /**
   * Execute a function within a "blocknote transaction".
   * All changes to the editor within the transaction will be grouped together, so that
   * we can dispatch them as a single operation (thus creating only a single undo step)
   *
   * @note There is no need to dispatch the transaction, as it will be automatically dispatched when the callback is complete.
   *
   * @example
   * ```ts
   * // All changes to the editor will be grouped together
   * editor.transact((tr) => {
   *   tr.insertText("Hello, world!");
   * // These two operations will be grouped together in a single undo step
   *   editor.transact((tr) => {
   *     tr.insertText("Hello, world!");
   *   });
   * });
   * ```
   */
  transact(e) {
    if (this.activeTransaction)
      return e(this.activeTransaction);
    try {
      this.activeTransaction = this.editor._tiptapEditor.state.tr;
      const t = e(this.activeTransaction), n = this.activeTransaction;
      return this.activeTransaction = null, n && // Only dispatch if the transaction was actually modified in some way
      (n.docChanged || n.selectionSet || n.scrolledIntoView || n.storedMarksSet || !n.isGeneric) && this.prosemirrorView.dispatch(n), t;
    } finally {
      this.activeTransaction = null;
    }
  }
  /**
   * Get the underlying prosemirror state
   * @note Prefer using `editor.transact` to read the current editor state, as that will ensure the state is up to date
   * @see https://prosemirror.net/docs/ref/#state.EditorState
   */
  get prosemirrorState() {
    if (this.activeTransaction)
      throw new Error(
        "`prosemirrorState` should not be called within a `transact` call, move the `prosemirrorState` call outside of the `transact` call or use `editor.transact` to read the current editor state"
      );
    return this.editor._tiptapEditor.state;
  }
  /**
   * Get the underlying prosemirror view
   * @see https://prosemirror.net/docs/ref/#view.EditorView
   */
  get prosemirrorView() {
    return this.editor._tiptapEditor.view;
  }
  isFocused() {
    var e;
    return ((e = this.prosemirrorView) == null ? void 0 : e.hasFocus()) || !1;
  }
  focus() {
    var e;
    (e = this.prosemirrorView) == null || e.focus();
  }
  /**
   * Checks if the editor is currently editable, or if it's locked.
   * @returns True if the editor is editable, false otherwise.
   */
  get isEditable() {
    if (!this.editor._tiptapEditor) {
      if (!this.editor.headless)
        throw new Error("no editor, but also not headless?");
      return !1;
    }
    return this.editor._tiptapEditor.isEditable === void 0 ? !0 : this.editor._tiptapEditor.isEditable;
  }
  /**
   * Makes the editor editable or locks it, depending on the argument passed.
   * @param editable True to make the editor editable, or false to lock it.
   */
  set isEditable(e) {
    if (!this.editor._tiptapEditor) {
      if (!this.editor.headless)
        throw new Error("no editor, but also not headless?");
      return;
    }
    this.editor._tiptapEditor.options.editable !== e && this.editor._tiptapEditor.setEditable(e);
  }
  /**
   * Undo the last action.
   */
  undo() {
    const e = this.editor.getExtension("yUndo");
    if (e)
      return this.exec(e.undoCommand);
    const t = this.editor.getExtension("history");
    if (t)
      return this.exec(t.undoCommand);
    throw new Error("No undo plugin found");
  }
  /**
   * Redo the last action.
   */
  redo() {
    const e = this.editor.getExtension("yUndo");
    if (e)
      return this.exec(e.redoCommand);
    const t = this.editor.getExtension("history");
    if (t)
      return this.exec(t.redoCommand);
    throw new Error("No redo plugin found");
  }
}
function Nn(o, e, t, n = { updateSelection: !0 }) {
  let { from: s, to: r } = typeof e == "number" ? { from: e, to: e } : { from: e.from, to: e.to }, i = !0, l = !0, a = "";
  if (t.forEach((c) => {
    c.check(), i && c.isText && c.marks.length === 0 ? a += c.text : i = !1, l = l ? c.isBlock : !1;
  }), s === r && l) {
    const { parent: c } = o.doc.resolve(s);
    c.isTextblock && !c.type.spec.code && !c.childCount && (s -= 1, r += 1);
  }
  return i ? o.insertText(a, s, r) : o.replaceWith(s, r, t), n.updateSelection && At(o, o.steps.length - 1, -1), !0;
}
class Ln {
  constructor(e) {
    this.editor = e;
  }
  /**
   * Insert a piece of content at the current cursor position.
   *
   * @param content can be a string, or array of partial inline content elements
   */
  insertInlineContent(e, { updateSelection: t = !1 } = {}) {
    const n = V(e, this.editor.pmSchema);
    this.editor.transact((s) => {
      Nn(
        s,
        {
          from: s.selection.from,
          to: s.selection.to
        },
        n,
        {
          updateSelection: t
        }
      );
    });
  }
  /**
   * Gets the active text styles at the text cursor position or at the end of the current selection if it's active.
   */
  getActiveStyles() {
    return this.editor.transact((e) => {
      const t = {}, n = e.selection.$to.marks();
      for (const s of n) {
        const r = this.editor.schema.styleSchema[s.type.name];
        if (!r) {
          // Links are not considered styles in blocknote
          s.type.name !== "link" && // "blocknoteIgnore" tagged marks (such as comments) are also not considered BlockNote "styles"
          !s.type.spec.blocknoteIgnore && console.warn("mark not found in styleschema", s.type.name);
          continue;
        }
        r.propSchema === "boolean" ? t[r.type] = !0 : t[r.type] = s.attrs.stringValue;
      }
      return t;
    });
  }
  /**
   * Adds styles to the currently selected content.
   * @param styles The styles to add.
   */
  addStyles(e) {
    for (const [t, n] of Object.entries(e)) {
      const s = this.editor.schema.styleSchema[t];
      if (!s)
        throw new Error(`style ${t} not found in styleSchema`);
      if (s.propSchema === "boolean")
        this.editor._tiptapEditor.commands.setMark(t);
      else if (s.propSchema === "string")
        this.editor._tiptapEditor.commands.setMark(t, {
          stringValue: n
        });
      else
        throw new U(s.propSchema);
    }
  }
  /**
   * Removes styles from the currently selected content.
   * @param styles The styles to remove.
   */
  removeStyles(e) {
    for (const t of Object.keys(e))
      this.editor._tiptapEditor.commands.unsetMark(t);
  }
  /**
   * Toggles styles on the currently selected content.
   * @param styles The styles to toggle.
   */
  toggleStyles(e) {
    for (const [t, n] of Object.entries(e)) {
      const s = this.editor.schema.styleSchema[t];
      if (!s)
        throw new Error(`style ${t} not found in styleSchema`);
      if (s.propSchema === "boolean")
        this.editor._tiptapEditor.commands.toggleMark(t);
      else if (s.propSchema === "string")
        this.editor._tiptapEditor.commands.toggleMark(t, {
          stringValue: n
        });
      else
        throw new U(s.propSchema);
    }
  }
  /**
   * Gets the currently selected text.
   */
  getSelectedText() {
    return this.editor.transact((e) => e.doc.textBetween(e.selection.from, e.selection.to));
  }
  /**
   * Gets the URL of the last link in the current selection, or `undefined` if there are no links in the selection.
   */
  getSelectedLinkUrl() {
    return this.editor._tiptapEditor.getAttributes("link").href;
  }
  /**
   * Creates a new link to replace the selected content.
   * @param url The link URL.
   * @param text The text to display the link with.
   */
  createLink(e, t) {
    if (e === "")
      return;
    const n = this.editor.pmSchema.mark("link", { href: e });
    this.editor.transact((s) => {
      const { from: r, to: i } = s.selection;
      t ? s.insertText(t, r, i).addMark(r, r + t.length, n) : s.setSelection(Qt.create(s.doc, i)).addMark(
        r,
        i,
        n
      );
    });
  }
}
function _n(o) {
  return Nt(o.state.selection.$from, (e) => e.type.name === "tableCell" || e.type.name === "tableHeader") !== void 0;
}
function Oe(o, e) {
  var s;
  const t = e.nodes.hardBreak;
  let n = R.empty;
  return o.forEach((r) => {
    r.isTextblock && r.childCount > 0 ? (n = n.append(r.content), n = n.addToEnd(t.create())) : r.isText ? n = n.addToEnd(r) : r.isBlock && r.childCount > 0 && (n = n.append(
      Oe(r.content, e)
    ), n = n.addToEnd(t.create()));
  }), ((s = n.lastChild) == null ? void 0 : s.type) === t && (n = n.cut(0, n.size - 1)), n;
}
function Dn(o, e) {
  const t = [];
  return o.forEach((n, s, r) => {
    r !== e && t.push(n);
  }), R.from(t);
}
function On(o, e) {
  const t = [];
  for (let n = 0; n < o.childCount; n++)
    if (o.child(n).type.name === "tableRow")
      if (t.length > 0 && t[t.length - 1].type.name === "table") {
        const s = t[t.length - 1], r = s.copy(s.content.addToEnd(o.child(n)));
        t[t.length - 1] = r;
      } else {
        const s = e.nodes.table.createChecked(
          void 0,
          o.child(n)
        );
        t.push(s);
      }
    else
      t.push(o.child(n));
  return o = R.from(t), o;
}
function $n(o, e) {
  let t = R.from(o.content);
  if (t = On(t, e.state.schema), _n(e)) {
    let n = !1;
    if (t.descendants((s) => {
      s.type.isInGroup("tableContent") && (n = !0);
    }), !n && // is the content valid for a table paragraph?
    !e.state.schema.nodes.tableParagraph.validContent(t))
      return new W(
        Oe(t, e.state.schema),
        0,
        0
      );
  }
  if (!Hn(t, e))
    return new W(t, o.openStart, o.openEnd);
  for (let n = 0; n < t.childCount; n++)
    if (t.child(n).type.spec.group === "blockContent") {
      const s = [t.child(n)];
      if (n + 1 < t.childCount && t.child(n + 1).type.name === "blockGroup") {
        const i = t.child(n + 1).child(0).child(0);
        (i.type.name === "bulletListItem" || i.type.name === "numberedListItem" || i.type.name === "checkListItem") && (s.push(t.child(n + 1)), t = Dn(t, n + 1));
      }
      const r = e.state.schema.nodes.blockContainer.createChecked(
        void 0,
        s
      );
      t = t.replaceChild(n, r);
    }
  return new W(t, o.openStart, o.openEnd);
}
function Hn(o, e) {
  var r, i;
  const t = o.childCount === 1, n = ((r = o.firstChild) == null ? void 0 : r.type.spec.content) === "inline*", s = ((i = o.firstChild) == null ? void 0 : i.type.spec.content) === "tableRow+";
  if (t) {
    if (n)
      return !1;
    if (s) {
      const l = B(e.state);
      if (l.isBlockContainer)
        return !(l.blockContent.node.type.spec.content === "tableRow+");
    }
  }
  return !0;
}
const Fn = {
  enableInputRules: !0,
  enablePasteRules: !0,
  enableCoreExtensions: !1
};
class $e extends Ce {
  constructor(t) {
    var c, d, u, p, h, f, m, b, g, x;
    super();
    /**
     * The underlying prosemirror schema
     */
    k(this, "pmSchema");
    k(this, "_tiptapEditor");
    /**
     * Used by React to store a reference to an `ElementRenderer` helper utility to make sure we can render React elements
     * in the correct context (used by `ReactRenderUtil`)
     */
    k(this, "elementRenderer", null);
    /**
     * Cache of all blocks. This makes sure we don't have to "recompute" blocks if underlying Prosemirror Nodes haven't changed.
     * This is especially useful when we want to keep track of the same block across multiple operations,
     * with this cache, blocks stay the same object reference (referential equality with ===).
     */
    k(this, "blockCache", /* @__PURE__ */ new WeakMap());
    /**
     * The dictionary contains translations for the editor.
     */
    k(this, "dictionary");
    /**
     * The schema of the editor. The schema defines which Blocks, InlineContent, and Styles are available in the editor.
     */
    k(this, "schema");
    k(this, "blockImplementations");
    k(this, "inlineContentImplementations");
    k(this, "styleImplementations");
    /**
     * The `uploadFile` method is what the editor uses when files need to be uploaded (for example when selecting an image to upload).
     * This method should set when creating the editor as this is application-specific.
     *
     * `undefined` means the application doesn't support file uploads.
     *
     * @param file The file that should be uploaded.
     * @returns The URL of the uploaded file OR an object containing props that should be set on the file block (such as an id)
     */
    k(this, "uploadFile");
    k(this, "onUploadStartCallbacks", []);
    k(this, "onUploadEndCallbacks", []);
    k(this, "resolveFileUrl");
    /**
     * Editor settings
     */
    k(this, "settings");
    // Manager instances
    k(this, "_blockManager");
    k(this, "_eventManager");
    k(this, "_exportManager");
    k(this, "_extensionManager");
    k(this, "_selectionManager");
    k(this, "_stateManager");
    k(this, "_styleManager");
    /**
     * Remove extension(s) from the editor
     */
    k(this, "unregisterExtension", (...t) => this._extensionManager.unregisterExtension(...t));
    /**
     * Register extension(s) to the editor
     */
    k(this, "registerExtension", (...t) => this._extensionManager.registerExtension(...t));
    /**
     * Get an extension from the editor
     */
    k(this, "getExtension", (...t) => this._extensionManager.getExtension(...t));
    /**
     * Mount the editor to a DOM element.
     *
     * @warning Not needed to call manually when using React, use BlockNoteView to take care of mounting
     */
    k(this, "mount", (t) => {
      this._tiptapEditor.mount({ mount: t });
    });
    /**
     * Unmount the editor from the DOM element it is bound to
     */
    k(this, "unmount", () => {
      this._tiptapEditor.unmount();
    });
    this.options = t, this.dictionary = t.dictionary || Ot, this.settings = {
      tables: {
        splitCells: ((c = t == null ? void 0 : t.tables) == null ? void 0 : c.splitCells) ?? !1,
        cellBackgroundColor: ((d = t == null ? void 0 : t.tables) == null ? void 0 : d.cellBackgroundColor) ?? !1,
        cellTextColor: ((u = t == null ? void 0 : t.tables) == null ? void 0 : u.cellTextColor) ?? !1,
        headers: ((p = t == null ? void 0 : t.tables) == null ? void 0 : p.headers) ?? !1
      }
    };
    const n = {
      defaultStyles: !0,
      schema: t.schema || wt.create(),
      ...t,
      placeholders: {
        ...this.dictionary.placeholders,
        ...t.placeholders
      }
    };
    if (this.schema = n.schema, this.blockImplementations = n.schema.blockSpecs, this.inlineContentImplementations = n.schema.inlineContentSpecs, this.styleImplementations = n.schema.styleSpecs, n.uploadFile) {
      const y = n.uploadFile;
      this.uploadFile = async (I, w) => {
        this.onUploadStartCallbacks.forEach(
          (A) => A.apply(this, [w])
        );
        try {
          return await y(I, w);
        } finally {
          this.onUploadEndCallbacks.forEach(
            (A) => A.apply(this, [w])
          );
        }
      };
    }
    this.resolveFileUrl = n.resolveFileUrl, this._eventManager = new Eo(this), this._extensionManager = new En(this, n);
    const s = this._extensionManager.getTiptapExtensions(), r = this._extensionManager.hasExtension("ySync") || this._extensionManager.hasExtension("liveblocksExtension");
    r && n.initialContent && console.warn(
      "When using Collaboration, initialContent might cause conflicts, because changes should come from the collaboration provider"
    );
    const i = {
      ...Fn,
      ...n._tiptapOptions,
      element: null,
      autofocus: n.autofocus ?? !1,
      extensions: s,
      editorProps: {
        ...(h = n._tiptapOptions) == null ? void 0 : h.editorProps,
        attributes: {
          // As of TipTap v2.5.0 the tabIndex is removed when the editor is not
          // editable, so you can't focus it. We want to revert this as we have
          // UI behaviour that relies on it.
          tabIndex: "0",
          ...(m = (f = n._tiptapOptions) == null ? void 0 : f.editorProps) == null ? void 0 : m.attributes,
          ...(b = n.domAttributes) == null ? void 0 : b.editor,
          class: ee(
            "bn-editor",
            n.defaultStyles ? "bn-default-styles" : "",
            ((x = (g = n.domAttributes) == null ? void 0 : g.editor) == null ? void 0 : x.class) || ""
          )
        },
        transformPasted: $n
      }
    };
    try {
      const y = n.initialContent || (r ? [
        {
          type: "paragraph",
          id: "initialBlockId"
        }
      ] : [
        {
          type: "paragraph",
          id: ke.options.generateID()
        }
      ]);
      if (!Array.isArray(y) || y.length === 0)
        throw new Error(
          "initialContent must be a non-empty array of blocks, received: " + y
        );
      const I = Lt(i.extensions), w = y.map(
        (He) => Z(He, I, this.schema.styleSchema).toJSON()
      ), A = _t(
        {
          type: "doc",
          content: [
            {
              type: "blockGroup",
              content: w
            }
          ]
        },
        I,
        i.parseOptions
      );
      this._tiptapEditor = new Dt({
        ...i,
        content: A.toJSON()
      }), this.pmSchema = this._tiptapEditor.schema;
    } catch (y) {
      throw new Error(
        "Error creating document from blocks passed as `initialContent`",
        { cause: y }
      );
    }
    let l;
    const a = this.pmSchema.nodes.doc.createAndFill;
    this.pmSchema.nodes.doc.createAndFill = (...y) => {
      if (l)
        return l;
      const I = a.apply(this.pmSchema.nodes.doc, y), w = JSON.parse(JSON.stringify(I.toJSON()));
      return w.content[0].content[0].attrs.id = "initialBlockId", l = ze.fromJSON(this.pmSchema, w), l;
    }, this.pmSchema.cached.blockNoteEditor = this, this._blockManager = new xo(this), this._exportManager = new $o(this), this._selectionManager = new In(this), this._stateManager = new An(this), this._styleManager = new Ln(this), this.emit("create");
  }
  static create(t) {
    return new $e(t ?? {});
  }
  /**
   * BlockNote extensions that are added to the editor, keyed by the extension key
   */
  get extensions() {
    return this._extensionManager.getExtensions();
  }
  /**
   * Execute a prosemirror command. This is mostly for backwards compatibility with older code.
   *
   * @note You should prefer the {@link transact} method when possible, as it will automatically handle the dispatching of the transaction and work across blocknote transactions.
   *
   * @example
   * ```ts
   * editor.exec((state, dispatch, view) => {
   *   dispatch(state.tr.insertText("Hello, world!"));
   * });
   * ```
   */
  exec(t) {
    return this._stateManager.exec(t);
  }
  /**
   * Check if a command can be executed. A command should return `false` if it is not valid in the current state.
   *
   * @example
   * ```ts
   * if (editor.canExec(command)) {
   *   // show button
   * } else {
   *   // hide button
   * }
   * ```
   */
  canExec(t) {
    return this._stateManager.canExec(t);
  }
  /**
   * Execute a function within a "blocknote transaction".
   * All changes to the editor within the transaction will be grouped together, so that
   * we can dispatch them as a single operation (thus creating only a single undo step)
   *
   * @note There is no need to dispatch the transaction, as it will be automatically dispatched when the callback is complete.
   *
   * @example
   * ```ts
   * // All changes to the editor will be grouped together
   * editor.transact((tr) => {
   *   tr.insertText("Hello, world!");
   * // These two operations will be grouped together in a single undo step
   *   editor.transact((tr) => {
   *     tr.insertText("Hello, world!");
   *   });
   * });
   * ```
   */
  transact(t) {
    return this._stateManager.transact(t);
  }
  /**
   * Get the underlying prosemirror state
   * @note Prefer using `editor.transact` to read the current editor state, as that will ensure the state is up to date
   * @see https://prosemirror.net/docs/ref/#state.EditorState
   */
  get prosemirrorState() {
    return this._stateManager.prosemirrorState;
  }
  /**
   * Get the underlying prosemirror view
   * @see https://prosemirror.net/docs/ref/#view.EditorView
   */
  get prosemirrorView() {
    return this._stateManager.prosemirrorView;
  }
  get domElement() {
    var t;
    if (!this.headless)
      return (t = this.prosemirrorView) == null ? void 0 : t.dom;
  }
  isFocused() {
    var t;
    return this.headless ? !1 : ((t = this.prosemirrorView) == null ? void 0 : t.hasFocus()) || !1;
  }
  get headless() {
    return !this._tiptapEditor.isInitialized;
  }
  /**
   * Focus on the editor
   */
  focus() {
    this.headless || this.prosemirrorView.focus();
  }
  /**
   * Blur the editor
   */
  blur() {
    var t;
    this.headless || (t = this.domElement) == null || t.blur();
  }
  // TODO move to extension
  onUploadStart(t) {
    return this.onUploadStartCallbacks.push(t), () => {
      const n = this.onUploadStartCallbacks.indexOf(t);
      n > -1 && this.onUploadStartCallbacks.splice(n, 1);
    };
  }
  onUploadEnd(t) {
    return this.onUploadEndCallbacks.push(t), () => {
      const n = this.onUploadEndCallbacks.indexOf(t);
      n > -1 && this.onUploadEndCallbacks.splice(n, 1);
    };
  }
  /**
   * @deprecated, use `editor.document` instead
   */
  get topLevelBlocks() {
    return this.document;
  }
  /**
   * Gets a snapshot of all top-level (non-nested) blocks in the editor.
   * @returns A snapshot of all top-level (non-nested) blocks in the editor.
   */
  get document() {
    return this._blockManager.document;
  }
  /**
   * Gets a snapshot of an existing block from the editor.
   * @param blockIdentifier The identifier of an existing block that should be
   * retrieved.
   * @returns The block that matches the identifier, or `undefined` if no
   * matching block was found.
   */
  getBlock(t) {
    return this._blockManager.getBlock(t);
  }
  /**
   * Gets a snapshot of the previous sibling of an existing block from the
   * editor.
   * @param blockIdentifier The identifier of an existing block for which the
   * previous sibling should be retrieved.
   * @returns The previous sibling of the block that matches the identifier.
   * `undefined` if no matching block was found, or it's the first child/block
   * in the document.
   */
  getPrevBlock(t) {
    return this._blockManager.getPrevBlock(t);
  }
  /**
   * Gets a snapshot of the next sibling of an existing block from the editor.
   * @param blockIdentifier The identifier of an existing block for which the
   * next sibling should be retrieved.
   * @returns The next sibling of the block that matches the identifier.
   * `undefined` if no matching block was found, or it's the last child/block in
   * the document.
   */
  getNextBlock(t) {
    return this._blockManager.getNextBlock(t);
  }
  /**
   * Gets a snapshot of the parent of an existing block from the editor.
   * @param blockIdentifier The identifier of an existing block for which the
   * parent should be retrieved.
   * @returns The parent of the block that matches the identifier. `undefined`
   * if no matching block was found, or the block isn't nested.
   */
  getParentBlock(t) {
    return this._blockManager.getParentBlock(t);
  }
  /**
   * Traverses all blocks in the editor depth-first, and executes a callback for each.
   * @param callback The callback to execute for each block. Returning `false` stops the traversal.
   * @param reverse Whether the blocks should be traversed in reverse order.
   */
  forEachBlock(t, n = !1) {
    this._blockManager.forEachBlock(t, n);
  }
  /**
   * Executes a callback whenever the editor's contents change.
   * @param callback The callback to execute.
   *
   * @deprecated use {@link BlockNoteEditor.onChange} instead
   */
  onEditorContentChange(t) {
    this._tiptapEditor.on("update", t);
  }
  /**
   * Executes a callback whenever the editor's selection changes.
   * @param callback The callback to execute.
   *
   * @deprecated use `onSelectionChange` instead
   */
  onEditorSelectionChange(t) {
    this._tiptapEditor.on("selectionUpdate", t);
  }
  /**
   * Executes a callback before any change is applied to the editor, allowing you to cancel the change.
   * @param callback The callback to execute.
   * @returns A function to remove the callback.
   */
  onBeforeChange(t) {
    return this._extensionManager.getExtension(ye).subscribe(t);
  }
  /**
   * Gets a snapshot of the current text cursor position.
   * @returns A snapshot of the current text cursor position.
   */
  getTextCursorPosition() {
    return this._selectionManager.getTextCursorPosition();
  }
  /**
   * Sets the text cursor position to the start or end of an existing block. Throws an error if the target block could
   * not be found.
   * @param targetBlock The identifier of an existing block that the text cursor should be moved to.
   * @param placement Whether the text cursor should be placed at the start or end of the block.
   */
  setTextCursorPosition(t, n = "start") {
    return this._selectionManager.setTextCursorPosition(t, n);
  }
  /**
   * Gets a snapshot of the current selection. This contains all blocks (included nested blocks)
   * that the selection spans across.
   *
   * If the selection starts / ends halfway through a block, the returned data will contain the entire block.
   */
  getSelection() {
    return this._selectionManager.getSelection();
  }
  /**
   * Gets a snapshot of the current selection. This contains all blocks (included nested blocks)
   * that the selection spans across.
   *
   * If the selection starts / ends halfway through a block, the returned block will be
   * only the part of the block that is included in the selection.
   */
  getSelectionCutBlocks(t = !1) {
    return this._selectionManager.getSelectionCutBlocks(t);
  }
  /**
   * Sets the selection to a range of blocks.
   * @param startBlock The identifier of the block that should be the start of the selection.
   * @param endBlock The identifier of the block that should be the end of the selection.
   */
  setSelection(t, n) {
    return this._selectionManager.setSelection(t, n);
  }
  /**
   * Checks if the editor is currently editable, or if it's locked.
   * @returns True if the editor is editable, false otherwise.
   */
  get isEditable() {
    return this._stateManager.isEditable;
  }
  /**
   * Makes the editor editable or locks it, depending on the argument passed.
   * @param editable True to make the editor editable, or false to lock it.
   */
  set isEditable(t) {
    this._stateManager.isEditable = t;
  }
  /**
   * Inserts new blocks into the editor. If a block's `id` is undefined, BlockNote generates one automatically. Throws an
   * error if the reference block could not be found.
   * @param blocksToInsert An array of partial blocks that should be inserted.
   * @param referenceBlock An identifier for an existing block, at which the new blocks should be inserted.
   * @param placement Whether the blocks should be inserted just before, just after, or nested inside the
   * `referenceBlock`.
   */
  insertBlocks(t, n, s = "before") {
    return this._blockManager.insertBlocks(
      t,
      n,
      s
    );
  }
  /**
   * Updates an existing block in the editor. Since updatedBlock is a PartialBlock object, some fields might not be
   * defined. These undefined fields are kept as-is from the existing block. Throws an error if the block to update could
   * not be found.
   * @param blockToUpdate The block that should be updated.
   * @param update A partial block which defines how the existing block should be changed.
   */
  updateBlock(t, n) {
    return this._blockManager.updateBlock(t, n);
  }
  /**
   * Removes existing blocks from the editor. Throws an error if any of the blocks could not be found.
   * @param blocksToRemove An array of identifiers for existing blocks that should be removed.
   */
  removeBlocks(t) {
    return this._blockManager.removeBlocks(t);
  }
  /**
   * Replaces existing blocks in the editor with new blocks. If the blocks that should be removed are not adjacent or
   * are at different nesting levels, `blocksToInsert` will be inserted at the position of the first block in
   * `blocksToRemove`. Throws an error if any of the blocks to remove could not be found.
   * @param blocksToRemove An array of blocks that should be replaced.
   * @param blocksToInsert An array of partial blocks to replace the old ones with.
   */
  replaceBlocks(t, n) {
    return this._blockManager.replaceBlocks(t, n);
  }
  /**
   * Undo the last action.
   */
  undo() {
    return this._stateManager.undo();
  }
  /**
   * Redo the last action.
   */
  redo() {
    return this._stateManager.redo();
  }
  /**
   * Insert a piece of content at the current cursor position.
   *
   * @param content can be a string, or array of partial inline content elements
   */
  insertInlineContent(t, { updateSelection: n = !1 } = {}) {
    this._styleManager.insertInlineContent(t, { updateSelection: n });
  }
  /**
   * Gets the active text styles at the text cursor position or at the end of the current selection if it's active.
   */
  getActiveStyles() {
    return this._styleManager.getActiveStyles();
  }
  /**
   * Adds styles to the currently selected content.
   * @param styles The styles to add.
   */
  addStyles(t) {
    this._styleManager.addStyles(t);
  }
  /**
   * Removes styles from the currently selected content.
   * @param styles The styles to remove.
   */
  removeStyles(t) {
    this._styleManager.removeStyles(t);
  }
  /**
   * Toggles styles on the currently selected content.
   * @param styles The styles to toggle.
   */
  toggleStyles(t) {
    this._styleManager.toggleStyles(t);
  }
  /**
   * Gets the currently selected text.
   */
  getSelectedText() {
    return this._styleManager.getSelectedText();
  }
  /**
   * Gets the URL of the last link in the current selection, or `undefined` if there are no links in the selection.
   */
  getSelectedLinkUrl() {
    return this._styleManager.getSelectedLinkUrl();
  }
  /**
   * Creates a new link to replace the selected content.
   * @param url The link URL.
   * @param text The text to display the link with.
   */
  createLink(t, n) {
    this._styleManager.createLink(t, n);
  }
  /**
   * Checks if the block containing the text cursor can be nested.
   */
  canNestBlock() {
    return this._blockManager.canNestBlock();
  }
  /**
   * Nests the block containing the text cursor into the block above it.
   */
  nestBlock() {
    this._blockManager.nestBlock();
  }
  /**
   * Checks if the block containing the text cursor is nested.
   */
  canUnnestBlock() {
    return this._blockManager.canUnnestBlock();
  }
  /**
   * Lifts the block containing the text cursor out of its parent.
   */
  unnestBlock() {
    this._blockManager.unnestBlock();
  }
  /**
   * Moves the selected blocks up. If the previous block has children, moves
   * them to the end of its children. If there is no previous block, but the
   * current blocks share a common parent, moves them out of & before it.
   */
  moveBlocksUp() {
    return this._blockManager.moveBlocksUp();
  }
  /**
   * Moves the selected blocks down. If the next block has children, moves
   * them to the start of its children. If there is no next block, but the
   * current blocks share a common parent, moves them out of & after it.
   */
  moveBlocksDown() {
    return this._blockManager.moveBlocksDown();
  }
  /**
   * Exports blocks into a simplified HTML string. To better conform to HTML standards, children of blocks which aren't list
   * items are un-nested in the output HTML.
   *
   * @param blocks An array of blocks that should be serialized into HTML.
   * @returns The blocks, serialized as an HTML string.
   */
  blocksToHTMLLossy(t = this.document) {
    return this._exportManager.blocksToHTMLLossy(t);
  }
  /**
   * Serializes blocks into an HTML string in the format that would normally be rendered by the editor.
   *
   * Use this method if you want to server-side render HTML (for example, a blog post that has been edited in BlockNote)
   * and serve it to users without loading the editor on the client (i.e.: displaying the blog post)
   *
   * @param blocks An array of blocks that should be serialized into HTML.
   * @returns The blocks, serialized as an HTML string.
   */
  blocksToFullHTML(t = this.document) {
    return this._exportManager.blocksToFullHTML(t);
  }
  /**
   * Parses blocks from an HTML string. Tries to create `Block` objects out of any HTML block-level elements, and
   * `InlineNode` objects from any HTML inline elements, though not all element types are recognized. If BlockNote
   * doesn't recognize an HTML element's tag, it will parse it as a paragraph or plain text.
   * @param html The HTML string to parse blocks from.
   * @returns The blocks parsed from the HTML string.
   */
  tryParseHTMLToBlocks(t) {
    return this._exportManager.tryParseHTMLToBlocks(t);
  }
  /**
   * Serializes blocks into a Markdown string. The output is simplified as Markdown does not support all features of
   * BlockNote - children of blocks which aren't list items are un-nested and certain styles are removed.
   * @param blocks An array of blocks that should be serialized into Markdown.
   * @returns The blocks, serialized as a Markdown string.
   */
  blocksToMarkdownLossy(t = this.document) {
    return this._exportManager.blocksToMarkdownLossy(t);
  }
  /**
   * Creates a list of blocks from a Markdown string. Tries to create `Block` and `InlineNode` objects based on
   * Markdown syntax, though not all symbols are recognized. If BlockNote doesn't recognize a symbol, it will parse it
   * as text.
   * @param markdown The Markdown string to parse blocks from.
   * @returns The blocks parsed from the Markdown string.
   */
  tryParseMarkdownToBlocks(t) {
    return this._exportManager.tryParseMarkdownToBlocks(t);
  }
  /**
   * A callback function that runs whenever the editor's contents change.
   *
   * @param callback The callback to execute.
   * @returns A function to remove the callback.
   */
  onChange(t, n) {
    return this._eventManager.onChange(t, n);
  }
  /**
   * A callback function that runs whenever the text cursor position or selection changes.
   *
   * @param callback The callback to execute.
   * @returns A function to remove the callback.
   */
  onSelectionChange(t, n) {
    return this._eventManager.onSelectionChange(
      t,
      n
    );
  }
  /**
   * A callback function that runs when the editor has been mounted.
   *
   * This can be useful for plugins to initialize themselves after the editor has been mounted.
   *
   * @param callback The callback to execute.
   * @returns A function to remove the callback.
   */
  onMount(t) {
    this._eventManager.onMount(t);
  }
  /**
   * A callback function that runs when the editor has been unmounted.
   *
   * This can be useful for plugins to clean up themselves after the editor has been unmounted.
   *
   * @param callback The callback to execute.
   * @returns A function to remove the callback.
   */
  onUnmount(t) {
    this._eventManager.onUnmount(t);
  }
  /**
   * Gets the bounding box of the current selection.
   * @returns The bounding box of the current selection.
   */
  getSelectionBoundingBox() {
    return this._selectionManager.getSelectionBoundingBox();
  }
  get isEmpty() {
    const t = this.document;
    return t.length === 0 || t.length === 1 && t[0].type === "paragraph" && t[0].content.length === 0;
  }
  /**
   * Paste HTML into the editor. Defaults to converting HTML to BlockNote HTML.
   * @param html The HTML to paste.
   * @param raw Whether to paste the HTML as is, or to convert it to BlockNote HTML.
   */
  pasteHTML(t, n = !1) {
    this._exportManager.pasteHTML(t, n);
  }
  /**
   * Paste text into the editor. Defaults to interpreting text as markdown.
   * @param text The text to paste.
   */
  pasteText(t) {
    return this._exportManager.pasteText(t);
  }
  /**
   * Paste markdown into the editor.
   * @param markdown The markdown to paste.
   */
  pasteMarkdown(t) {
    return this._exportManager.pasteMarkdown(t);
  }
}
class fs {
  constructor(e, t, n) {
    this.mappings = t, this.options = n;
  }
  async resolveFile(e) {
    var n;
    if (!((n = this.options) != null && n.resolveFileUrl))
      return (await fetch(e)).blob();
    const t = await this.options.resolveFileUrl(e);
    return t instanceof Blob ? t : (await fetch(t)).blob();
  }
  mapStyles(e) {
    return Object.entries(e).map(([n, s]) => this.mappings.styleMapping[n](s, this));
  }
  mapInlineContent(e) {
    return this.mappings.inlineContentMapping[e.type](
      e,
      this
    );
  }
  transformInlineContent(e) {
    return e.map((t) => this.mapInlineContent(t));
  }
  async mapBlock(e, t, n, s) {
    return this.mappings.blockMapping[e.type](
      e,
      this,
      t,
      n,
      s
    );
  }
}
function hs(o) {
  return {
    createBlockMapping: (e) => e,
    createInlineContentMapping: (e) => e,
    createStyleMapping: (e) => e
  };
}
function ms(o, ...e) {
  const t = [...o];
  for (const n of e)
    for (const s of n) {
      const r = t.findLastIndex(
        (i) => i.group === s.group
      );
      r === -1 ? t.push(s) : t.splice(r + 1, 0, s);
    }
  return t;
}
export {
  ye as BlockChangeExtension,
  $e as BlockNoteEditor,
  wt as BlockNoteSchema,
  Ls as COLORS_DARK_MODE_DEFAULT,
  _s as COLORS_DEFAULT,
  Si as CustomBlockNoteSchema,
  Pt as DEFAULT_LINK_PROTOCOL,
  mi as DRAG_EXCLUSION_CLASSNAME,
  mt as DropCursorExtension,
  Ds as EMPTY_CELL_HEIGHT,
  Xe as EMPTY_CELL_WIDTH,
  Ce as EventEmitter,
  fs as Exporter,
  Os as FILE_AUDIO_ICON_SVG,
  $s as FILE_IMAGE_ICON_SVG,
  Hs as FILE_VIDEO_ICON_SVG,
  J as FilePanelExtension,
  dt as ForkYDocExtension,
  Y as FormattingToolbarExtension,
  ve as HTMLToBlocks,
  Ct as HistoryExtension,
  kt as LinkToolbarExtension,
  bt as NodeSelectionKeyboardExtension,
  gt as PlaceholderExtension,
  xt as PreviousBlockTypeExtension,
  ht as SchemaMigration,
  Wt as ShowSelectionExtension,
  Bt as SideMenuExtension,
  ki as SideMenuView,
  ge as SuggestionMenu,
  St as TableHandlesExtension,
  bi as TableHandlesView,
  yt as TrailingNodeExtension,
  ke as UniqueID,
  U as UnreachableCaseError,
  Et as VALID_LINK_PROTOCOLS,
  ut as YCursorExtension,
  pt as YSyncExtension,
  ft as YUndoExtension,
  Fs as addDefaultPropsExternalHTML,
  G as addInlineContentAttributes,
  Je as addInlineContentKeyboardShortcuts,
  Vs as addNodeAndExtensionsToSpec,
  Us as addStyleAttributes,
  zs as applyNonSelectableBlockFix,
  gs as assertEmpty,
  Rs as audioParse,
  Gs as audioRender,
  Ws as audioToExternalHTML,
  js as blockHasType,
  Z as blockToNode,
  at as blocksToMarkdown,
  qs as camelToDataKebab,
  Ks as captureCellAnchor,
  xi as checkPageBreakBlocksInSchema,
  ct as cleanHTMLToMarkdown,
  ms as combineByGroup,
  qe as contentNodeToInlineContent,
  je as contentNodeToTableContent,
  Js as createAudioBlockConfig,
  Ys as createAudioBlockSpec,
  Qs as createBlockConfig,
  Xs as createBlockSpec,
  Zs as createBlockSpecFromTiptapNode,
  er as createBulletListItemBlockConfig,
  tr as createBulletListItemBlockSpec,
  or as createCheckListItemBlockSpec,
  nr as createCheckListItemConfig,
  sr as createCodeBlockConfig,
  rr as createCodeBlockSpec,
  ir as createDefaultBlockDOMOutputSpec,
  ar as createDividerBlockConfig,
  cr as createDividerBlockSpec,
  Vt as createExtension,
  Be as createExternalHTMLExporter,
  lr as createFileBlockConfig,
  dr as createFileBlockSpec,
  ur as createHeadingBlockConfig,
  pr as createHeadingBlockSpec,
  fr as createImageBlockConfig,
  hr as createImageBlockSpec,
  ps as createInlineContentSpec,
  Qe as createInlineContentSpecFromTipTapNode,
  lo as createInternalHTMLSerializer,
  mr as createInternalInlineContentSpec,
  kr as createInternalStyleSpec,
  br as createNumberedListItemBlockConfig,
  gr as createNumberedListItemBlockSpec,
  Ei as createPageBreakBlockConfig,
  Pi as createPageBreakBlockSpec,
  Br as createParagraphBlockConfig,
  yr as createParagraphBlockSpec,
  Cr as createQuoteBlockConfig,
  Sr as createQuoteBlockSpec,
  Ii as createStore,
  xr as createStyleSpec,
  Er as createStyleSpecFromTipTapMark,
  Pr as createTableBlockSpec,
  Tr as createToggleListItemBlockConfig,
  wr as createToggleListItemBlockSpec,
  Mr as createToggleWrapper,
  vr as createVideoBlockConfig,
  Ir as createVideoBlockSpec,
  Ar as defaultBlockSpecs,
  Nr as defaultBlockToHTML,
  Lr as defaultInlineContentSchema,
  _r as defaultInlineContentSpecs,
  Dr as defaultProps,
  Or as defaultStyleSchema,
  $r as defaultStyleSpecs,
  Hr as defaultToggledState,
  We as docToBlocks,
  Fr as editorHasBlockWithType,
  Pn as expandPMRangeToWords,
  Vr as fileParse,
  Ur as filenameFromURL,
  zr as filterSuggestionItems,
  H as fixColumnList,
  Rr as formatKeyboardShortcut,
  tt as getBackgroundColorAttribute,
  Bo as getBlock,
  Bs as getBlockCache,
  Gr as getBlockFromPos,
  K as getBlockInfo,
  v as getBlockInfoFromResolvedPos,
  B as getBlockInfoFromSelection,
  z as getBlockInfoFromTransaction,
  ys as getBlockInfoWithManualOffset,
  be as getBlockNoteSchema,
  Cs as getBlockSchema,
  it as getBlocksChangedByTransaction,
  Ss as getColspan,
  gi as getDefaultEmojiPickerItems,
  Wr as getDefaultSlashMenuItems,
  Xt as getInlineContentParseRules,
  xs as getInlineContentSchema,
  jr as getInlineContentSchemaFromSpecs,
  qr as getLanguageId,
  D as getNearestBlockPos,
  Co as getNextBlock,
  P as getNodeById,
  Ti as getPageBreakSlashMenuItems,
  So as getParentBlock,
  Kr as getParseRules,
  E as getPmSchema,
  yo as getPrevBlock,
  Es as getRowspan,
  Jr as getStyleParseRules,
  Ps as getStyleSchema,
  Yr as getStyleSchemaFromSpecs,
  Qr as getTextAlignmentAttribute,
  st as getTextColorAttribute,
  Xr as imageParse,
  Zr as imageRender,
  ei as imageToExternalHTML,
  V as inlineContentToNodes,
  Zt as insertBlocks,
  ti as insertOrUpdateBlockForSlashMenu,
  oi as isAppleOS,
  Q as isEmptyColumn,
  Ts as isLinkInlineContent,
  ni as isNodeBlock,
  ws as isPartialLinkInlineContent,
  Ms as isPartialTableCell,
  si as isSafari,
  vs as isStyledTextInlineContent,
  Is as isTableCell,
  ri as isTableCellSelection,
  et as isVideoUrl,
  As as mapTableCell,
  hs as mappingFactory,
  Oo as markdownToBlocks,
  Ie as markdownToHTML,
  ee as mergeCSSClasses,
  ii as mergeParagraphs,
  C as nodeToBlock,
  q as nodeToCustomInlineContent,
  ai as parseAudioElement,
  ci as parseDefaultProps,
  Ye as propsToAttributes,
  Ke as prosemirrorSliceToSlicedBlocks,
  ae as removeAndInsertBlocks,
  eo as removeEmptyColumns,
  Ne as selectedFragmentToHTML,
  Bi as sideMenuPluginKey,
  li as stylePropsToAttributes,
  Ge as tableContentToNodes,
  yi as tableHandlesPluginKey,
  di as tablePropSchema,
  ui as trackPosition,
  Ze as updateBlock,
  nt as updateBlockCommand,
  rt as updateBlockTr,
  wi as uploadToTmpFilesDotOrg_DEV_ONLY,
  pi as videoParse,
  Mi as withPageBreak,
  fi as wrapInBlockStructure
};
