// @blocknote/core/dist/blocks.js
import { E as t, a as c, F as s, r as o, N as l, a4 as r, b as i, d as g, e as k, a0 as B, c as n, f as p, x as d, y as u, A as S, z as f, g as C, h as m, T as I, j as T, k as L, l as P, n as E, o as _, q as h, s as D, w as H, B as b, C as A, H as M, I as O, J as x, K as N, M as V, D as F, G, S as y, O as v, Q as W, W as Y, U as w, _ as Q, Z as R, a2 as U, Y as j, X as q, R as z, $ as J, m as K, a5 as X, i as Z, a7 as $, a6 as ee, t as ae, u as te, v as ce, a1 as se, V as oe, p as le, a3 as re, L as ie, P as ge } from "./defaultBlocks-ZzGbYgQn.js";
import { B as Be, b as ne, c as pe, a as de, g as ue, u as Se, w as fe } from "./BlockNoteSchema-ooiKsd5B.js";
export {
  Be as BlockNoteSchema,
  t as EMPTY_CELL_HEIGHT,
  c as EMPTY_CELL_WIDTH,
  s as FILE_AUDIO_ICON_SVG,
  o as FILE_IMAGE_ICON_SVG,
  l as FILE_VIDEO_ICON_SVG,
  r as addDefaultPropsExternalHTML,
  i as audioParse,
  g as audioRender,
  k as audioToExternalHTML,
  B as blockHasType,
  ne as checkPageBreakBlocksInSchema,
  n as createAudioBlockConfig,
  p as createAudioBlockSpec,
  d as createBulletListItemBlockConfig,
  u as createBulletListItemBlockSpec,
  S as createCheckListItemBlockSpec,
  f as createCheckListItemConfig,
  C as createCodeBlockConfig,
  m as createCodeBlockSpec,
  I as createDefaultBlockDOMOutputSpec,
  T as createDividerBlockConfig,
  L as createDividerBlockSpec,
  P as createFileBlockConfig,
  E as createFileBlockSpec,
  _ as createHeadingBlockConfig,
  h as createHeadingBlockSpec,
  D as createImageBlockConfig,
  H as createImageBlockSpec,
  b as createNumberedListItemBlockConfig,
  A as createNumberedListItemBlockSpec,
  pe as createPageBreakBlockConfig,
  de as createPageBreakBlockSpec,
  M as createParagraphBlockConfig,
  O as createParagraphBlockSpec,
  x as createQuoteBlockConfig,
  N as createQuoteBlockSpec,
  V as createTableBlockSpec,
  F as createToggleListItemBlockConfig,
  G as createToggleListItemBlockSpec,
  y as createToggleWrapper,
  v as createVideoBlockConfig,
  W as createVideoBlockSpec,
  Y as defaultBlockSpecs,
  w as defaultBlockToHTML,
  Q as defaultInlineContentSchema,
  R as defaultInlineContentSpecs,
  U as defaultProps,
  j as defaultStyleSchema,
  q as defaultStyleSpecs,
  z as defaultToggledState,
  J as editorHasBlockWithType,
  K as fileParse,
  X as getBackgroundColorAttribute,
  Z as getLanguageId,
  ue as getPageBreakSlashMenuItems,
  $ as getTextAlignmentAttribute,
  ee as getTextColorAttribute,
  ae as imageParse,
  te as imageRender,
  ce as imageToExternalHTML,
  se as isTableCellSelection,
  oe as mergeParagraphs,
  le as parseAudioElement,
  re as parseDefaultProps,
  ie as tablePropSchema,
  Se as uploadToTmpFilesDotOrg_DEV_ONLY,
  ge as videoParse,
  fe as withPageBreak
};
