// @blocknote/core/dist/extensions.js
import { B as a, e as n, D as o, c as i, F as t, d as r, H as l, L as E, N as x, P as S, f as u, S as d, h as g, g as c, j as L, T, k as m, V as P, Y as b, a as D, b as M, i as O, s as f, t as h } from "./TrailingNode-GzE59m_7.js";
import { a8 as C, a9 as I, ac as N, ab as p, aa as y } from "./defaultBlocks-ZzGbYgQn.js";
import { S as _ } from "./ShowSelection-B0ch3unP.js";
export {
  a as BlockChangeExtension,
  n as DEFAULT_LINK_PROTOCOL,
  o as DRAG_EXCLUSION_CLASSNAME,
  i as DropCursorExtension,
  C as FilePanelExtension,
  t as ForkYDocExtension,
  r as FormattingToolbarExtension,
  l as HistoryExtension,
  E as LinkToolbarExtension,
  x as NodeSelectionKeyboardExtension,
  S as PlaceholderExtension,
  u as PreviousBlockTypeExtension,
  d as SchemaMigration,
  _ as ShowSelectionExtension,
  g as SideMenuExtension,
  c as SideMenuView,
  I as SuggestionMenu,
  L as TableHandlesExtension,
  T as TableHandlesView,
  m as TrailingNodeExtension,
  P as VALID_LINK_PROTOCOLS,
  b as YCursorExtension,
  D as YSyncExtension,
  M as YUndoExtension,
  N as filterSuggestionItems,
  O as getDefaultEmojiPickerItems,
  p as getDefaultSlashMenuItems,
  y as insertOrUpdateBlockForSlashMenu,
  f as sideMenuPluginKey,
  h as tableHandlesPluginKey
};
