// @blocknote/mantine/dist/blocknote-mantine.js
import { jsx as s, jsxs as v } from "react/jsx-runtime";
import { assertEmpty as u, isSafari as R, mergeCSSClasses as B, COLORS_DEFAULT as se, COLORS_DARK_MODE_DEFAULT as ae } from "@blocknote/core";
import { mergeRefs as ie, useDictionary as le, useBlockNoteContext as V, FormattingToolbarController as ce, getFormattingToolbarItems as de, FormattingToolbar as ue, elementOverflow as K, usePrefersColorScheme as me, ComponentsContext as ge, BlockNoteViewRaw as fe } from "@blocknote/react";
import { Button as T, ActionIcon as E, Tooltip as U, Stack as L, Text as C, Group as b, Chip as he, Divider as be, Card as W, Skeleton as $, Avatar as pe, TextInput as Z, Menu as p, CheckIcon as z, Tabs as P, LoadingOverlay as ve, FileInput as Se, PopoverDropdown as ye, PopoverTarget as we, Popover as xe, Badge as Ce, Loader as q, Flex as Ne, MantineContext as Te, MantineProvider as ke } from "@mantine/core";
import N, { forwardRef as m, useState as J, useContext as Q, createContext as Pe, useRef as X, useEffect as Y, useCallback as Fe } from "react";
import { useHover as Me, mergeRefs as j, useFocusWithin as Oe, useFocusTrap as De } from "@mantine/hooks";
const ee = (e, n, t = !1) => {
  const o = [];
  function r(a, i = "--bn") {
    for (const l in a) {
      const c = l.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase(), d = `${i}-${c}`;
      typeof a[l] != "object" ? (typeof a[l] == "number" && (a[l] = `${a[l]}px`), t ? n.style.removeProperty(d) : n.style.setProperty(d, a[l].toString())) : r(a[l], d);
    }
  }
  return r(e), o;
}, A = (e, n) => ee(e, n), Ie = {
  colors: {
    editor: {
      text: void 0,
      background: void 0
    },
    menu: {
      text: void 0,
      background: void 0
    },
    tooltip: {
      text: void 0,
      background: void 0
    },
    hovered: {
      text: void 0,
      background: void 0
    },
    selected: {
      text: void 0,
      background: void 0
    },
    disabled: {
      text: void 0,
      background: void 0
    },
    shadow: void 0,
    border: void 0,
    sideMenu: void 0,
    highlights: {
      gray: {
        text: void 0,
        background: void 0
      },
      brown: {
        text: void 0,
        background: void 0
      },
      red: {
        text: void 0,
        background: void 0
      },
      orange: {
        text: void 0,
        background: void 0
      },
      yellow: {
        text: void 0,
        background: void 0
      },
      green: {
        text: void 0,
        background: void 0
      },
      blue: {
        text: void 0,
        background: void 0
      },
      purple: {
        text: void 0,
        background: void 0
      },
      pink: {
        text: void 0,
        background: void 0
      }
    }
  },
  borderRadius: void 0,
  fontFamily: void 0
}, Re = (e) => ee(Ie, e, !0), te = (e) => /* @__PURE__ */ v(L, { gap: 0, className: "bn-tooltip", children: [
  /* @__PURE__ */ s(C, { size: "sm", lineClamp: 5, children: e.mainTooltip }),
  e.secondaryTooltip && /* @__PURE__ */ s(C, { size: "xs", lineClamp: 5, children: e.secondaryTooltip })
] }), O = m(
  (e, n) => {
    const {
      className: t,
      children: o,
      mainTooltip: r,
      secondaryTooltip: a,
      icon: i,
      isSelected: l,
      isDisabled: c,
      onClick: d,
      label: f,
      variant: h,
      ...S
    } = e;
    u(S, !1);
    const [x, y] = J(!1), k = o ? /* @__PURE__ */ s(
      T,
      {
        "aria-label": f,
        className: t,
        onMouseDown: (w) => {
          R() && w.currentTarget.focus();
        },
        onClick: (w) => {
          y(!0), d == null || d(w);
        },
        onPointerLeave: () => y(!1),
        "aria-pressed": l,
        "data-selected": l || void 0,
        "data-test": r ? r.slice(0, 1).toLowerCase() + r.replace(/\s+/g, "").slice(1) : void 0,
        size: h === "compact" ? "compact-xs" : "xs",
        disabled: c || !1,
        ref: n,
        ...S,
        children: o
      }
    ) : /* @__PURE__ */ s(
      E,
      {
        className: t,
        "aria-label": f,
        onMouseDown: (w) => {
          R() && w.currentTarget.focus();
        },
        onClick: (w) => {
          y(!0), d == null || d(w);
        },
        onPointerLeave: () => y(!1),
        "aria-pressed": l,
        "data-selected": l || void 0,
        "data-test": r ? r.slice(0, 1).toLowerCase() + r.replace(/\s+/g, "").slice(1) : void 0,
        size: h === "compact" ? 20 : 30,
        disabled: c || !1,
        ref: n,
        ...S,
        children: i
      }
    );
    return r ? /* @__PURE__ */ s(
      U,
      {
        disabled: x,
        withinPortal: !1,
        label: /* @__PURE__ */ s(
          te,
          {
            mainTooltip: r,
            secondaryTooltip: a
          }
        ),
        children: k
      }
    ) : k;
  }
), ze = m((e, n) => {
  const {
    className: t,
    text: o,
    icon: r,
    isSelected: a,
    mainTooltip: i,
    secondaryTooltip: l,
    onClick: c,
    onMouseEnter: d,
    ...f
  } = e;
  u(f, !1);
  const [h, S] = J(!1), x = /* @__PURE__ */ v(
    he,
    {
      className: t,
      checked: a === !0,
      wrapperProps: {
        onMouseEnter: d,
        onMouseLeave: () => S(!1),
        onClick: (y) => {
          S(!0), c == null || c(y);
        }
      },
      variant: "light",
      icon: null,
      ref: n,
      children: [
        /* @__PURE__ */ s("span", { children: r }),
        /* @__PURE__ */ s("span", { children: o })
      ]
    }
  );
  return !i || h ? x : /* @__PURE__ */ s(
    U,
    {
      refProp: "rootRef",
      withinPortal: !1,
      label: /* @__PURE__ */ s(
        te,
        {
          mainTooltip: i,
          secondaryTooltip: l
        }
      ),
      children: x
    }
  );
}), Be = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r), /* @__PURE__ */ s(b, { className: t, ref: n, children: o });
}), Ee = m((e, n) => {
  const {
    className: t,
    children: o,
    headerText: r,
    selected: a,
    onFocus: i,
    onBlur: l,
    tabIndex: c,
    ...d
  } = e;
  return u(d, !1), /* @__PURE__ */ v(
    W,
    {
      className: B(t, a ? "selected" : ""),
      onFocus: i,
      onBlur: l,
      tabIndex: c,
      ref: n,
      children: [
        r && /* @__PURE__ */ s(C, { className: "bn-header-text", children: r }),
        o
      ]
    }
  );
}), Le = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r, !1), /* @__PURE__ */ s(W.Section, { className: t, ref: n, children: o });
}), je = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r, !1), /* @__PURE__ */ s(
    be,
    {
      className: t,
      label: /* @__PURE__ */ s(C, { children: o }),
      ref: n
    }
  );
}), $e = m((e, n) => {
  const { authorInfo: t, timeString: o, edited: r, ...a } = e, i = le();
  return u(a, !1), t === "loading" ? /* @__PURE__ */ v(b, { children: [
    /* @__PURE__ */ s($, { height: 24, width: 24 }),
    /* @__PURE__ */ s("div", { children: /* @__PURE__ */ s($, { height: 12, width: 100 }) })
  ] }) : /* @__PURE__ */ v(b, { children: [
    /* @__PURE__ */ s(
      pe,
      {
        src: t.avatarUrl,
        alt: t.username,
        radius: "xl",
        size: "sm",
        color: "initials"
      }
    ),
    /* @__PURE__ */ v(C, { fz: "sm", fw: "bold", children: [
      t.username,
      /* @__PURE__ */ v(C, { fz: "xs", c: "dimmed", span: !0, ml: "xs", children: [
        o,
        " ",
        r && `(${i.comments.edited})`
      ] })
    ] })
  ] });
}), Ae = m((e, n) => {
  const {
    className: t,
    showActions: o,
    authorInfo: r,
    timeString: a,
    edited: i,
    actions: l,
    emojiPickerOpen: c,
    children: d,
    ...f
  } = e, { hovered: h, ref: S } = Me(), x = ie([n, S]);
  return u(f, !1), /* @__PURE__ */ v(b, { pos: "relative", ref: x, className: t, children: [
    l && (o === !0 || o === void 0 || o === "hover" && h || c) ? /* @__PURE__ */ s(
      b,
      {
        style: {
          position: "absolute",
          right: 0,
          top: 0,
          zIndex: 10
        },
        children: l
      }
    ) : null,
    /* @__PURE__ */ s($e, { ...e }),
    d
  ] });
}), _e = m((e, n) => {
  const { className: t, autoFocus: o, onFocus: r, onBlur: a, editor: i, editable: l, ...c } = e;
  u(c, !1);
  const d = V();
  return /* @__PURE__ */ s(
    Mt,
    {
      autoFocus: o,
      className: t,
      editor: e.editor,
      sideMenu: !1,
      slashMenu: !1,
      tableHandles: !1,
      filePanel: !1,
      formattingToolbar: !1,
      editable: l,
      theme: d == null ? void 0 : d.colorSchemePreference,
      ref: n,
      onFocus: r,
      onBlur: a,
      children: /* @__PURE__ */ s(
        ce,
        {
          formattingToolbar: Ge
        }
      )
    }
  );
}), Ge = () => {
  const e = de([]).filter(
    (n) => n.key !== "nestBlockButton" && n.key !== "unnestBlockButton"
  );
  return /* @__PURE__ */ s(ue, { blockTypeSelectItems: [], children: e });
}, He = m((e, n) => {
  const {
    className: t,
    name: o,
    label: r,
    variant: a,
    icon: i,
    value: l,
    autoFocus: c,
    placeholder: d,
    disabled: f,
    onKeyDown: h,
    onChange: S,
    onSubmit: x,
    autoComplete: y,
    "aria-activedescendant": k,
    rightSection: w,
    ...re
  } = e;
  return u(re), /* @__PURE__ */ s(
    Z,
    {
      size: "xs",
      className: B(
        t || "",
        a === "large" ? "bn-mt-input-large" : ""
      ),
      ref: n,
      name: o,
      label: r,
      leftSection: i,
      value: l,
      autoFocus: c,
      "data-autofocus": c ? "true" : void 0,
      rightSection: w,
      placeholder: d,
      disabled: f,
      onKeyDown: h,
      onChange: S,
      onSubmit: x,
      autoComplete: y,
      "aria-activedescendant": k
    }
  );
}), Ve = Pe(void 0), Ke = (e) => {
  const { children: n, onOpenChange: t, position: o, sub: r, ...a } = e;
  return u(a), r ? /* @__PURE__ */ s(
    p.Sub,
    {
      transitionProps: { duration: 250, exitDelay: 250 },
      withinPortal: !1,
      middlewares: { flip: !0, shift: !0, inline: !1, size: !0 },
      onChange: t,
      position: o,
      children: n
    }
  ) : /* @__PURE__ */ s(
    p,
    {
      withinPortal: !1,
      middlewares: { flip: !0, shift: !0, inline: !1, size: !0 },
      onChange: t,
      position: o,
      returnFocus: !1,
      children: n
    }
  );
}, Ue = m((e, n) => {
  const { className: t, children: o, icon: r, checked: a, subTrigger: i, onClick: l, ...c } = e;
  return u(c, !1), i ? /* @__PURE__ */ s(
    p.Sub.Item,
    {
      className: t,
      ref: n,
      leftSection: r,
      rightSection: a ? /* @__PURE__ */ s(z, { size: 10 }) : a === !1 ? /* @__PURE__ */ s("div", { className: "bn-tick-space" }) : null,
      onClick: l,
      ...c,
      children: o
    }
  ) : /* @__PURE__ */ s(
    p.Item,
    {
      className: t,
      ref: n,
      leftSection: r,
      rightSection: a ? /* @__PURE__ */ s(z, { size: 10 }) : a === !1 ? /* @__PURE__ */ s("div", { className: "bn-tick-space" }) : null,
      onClick: l,
      ...c,
      children: o
    }
  );
}), We = (e) => {
  const {
    children: n,
    sub: t,
    // unused
    ...o
  } = e;
  return u(o), t ? /* @__PURE__ */ s(p.Sub.Target, { children: n }) : /* @__PURE__ */ s(p.Target, { children: n });
}, Ze = m((e, n) => {
  const {
    className: t,
    children: o,
    sub: r,
    //unused
    ...a
  } = e;
  u(a);
  const i = Q(Ve);
  return r ? /* @__PURE__ */ s(
    p.Sub.Dropdown,
    {
      className: t,
      ref: n,
      onMouseOver: i == null ? void 0 : i.onMenuMouseOver,
      onMouseLeave: i == null ? void 0 : i.onMenuMouseLeave,
      children: o
    }
  ) : /* @__PURE__ */ s(
    p.Dropdown,
    {
      className: t,
      ref: n,
      onMouseOver: i == null ? void 0 : i.onMenuMouseOver,
      onMouseLeave: i == null ? void 0 : i.onMenuMouseLeave,
      children: o
    }
  );
}), qe = m((e, n) => {
  const { className: t, ...o } = e;
  return u(o), /* @__PURE__ */ s(p.Divider, { className: t, ref: n });
}), Je = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r), /* @__PURE__ */ s(p.Label, { className: t, ref: n, children: o });
}), Qe = m((e, n) => {
  const {
    className: t,
    children: o,
    icon: r,
    onClick: a,
    onDragEnd: i,
    onDragStart: l,
    draggable: c,
    label: d,
    ...f
  } = e;
  return u(f, !1), r ? /* @__PURE__ */ s(
    E,
    {
      size: 24,
      className: t,
      ref: n,
      onClick: a,
      onDragEnd: i,
      onDragStart: l,
      draggable: c,
      "aria-label": d,
      ...f,
      children: r
    }
  ) : /* @__PURE__ */ s(
    T,
    {
      className: t,
      ref: n,
      onClick: a,
      onDragEnd: i,
      onDragStart: l,
      draggable: c,
      "aria-label": d,
      ...f,
      children: o
    }
  );
}), Xe = m((e, n) => {
  const {
    className: t,
    tabs: o,
    defaultOpenTab: r,
    openTab: a,
    setOpenTab: i,
    loading: l,
    ...c
  } = e;
  return u(c), /* @__PURE__ */ s(b, { className: t, ref: n, children: /* @__PURE__ */ v(
    P,
    {
      value: a,
      defaultValue: r,
      onChange: i,
      children: [
        l && /* @__PURE__ */ s(ve, { visible: l }),
        /* @__PURE__ */ s(P.List, { children: o.map((d) => /* @__PURE__ */ s(
          P.Tab,
          {
            "data-test": `${d.name.toLowerCase()}-tab`,
            value: d.name,
            children: d.name
          },
          d.name
        )) }),
        o.map((d) => /* @__PURE__ */ s(P.Panel, { value: d.name, children: d.tabPanel }, d.name))
      ]
    }
  ) });
}), Ye = m((e, n) => {
  const { className: t, children: o, onClick: r, label: a, ...i } = e;
  return u(i), /* @__PURE__ */ s(
    T,
    {
      size: "xs",
      "aria-label": a,
      className: t,
      ref: n,
      onClick: r,
      ...i,
      children: o
    }
  );
}), et = m((e, n) => {
  const { className: t, accept: o, value: r, placeholder: a, onChange: i, ...l } = e;
  return u(l), /* @__PURE__ */ s(
    Se,
    {
      size: "xs",
      className: t,
      ref: n,
      accept: o,
      value: r,
      placeholder: a,
      onChange: i,
      ...l
    }
  );
}), tt = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r), /* @__PURE__ */ s("div", { className: t, ref: n, children: o });
}), nt = m((e, n) => {
  const { className: t, value: o, placeholder: r, onKeyDown: a, onChange: i, ...l } = e;
  return u(l), /* @__PURE__ */ s(
    Z,
    {
      size: "xs",
      "data-test": "embed-input",
      className: t,
      ref: n,
      value: o,
      placeholder: r,
      onKeyDown: a,
      onChange: i
    }
  );
}), ot = (e) => {
  const { open: n, onOpenChange: t, position: o, children: r, ...a } = e;
  return u(a), /* @__PURE__ */ s(
    xe,
    {
      middlewares: { size: { padding: 20 } },
      withinPortal: !1,
      opened: n,
      onChange: t,
      position: o,
      zIndex: 1e4,
      children: r
    }
  );
}, rt = (e) => {
  const { children: n, ...t } = e;
  return u(t), /* @__PURE__ */ s(we, { children: n });
}, st = m((e, n) => {
  const {
    className: t,
    children: o,
    variant: r,
    // unused
    ...a
  } = e;
  return u(a), /* @__PURE__ */ s(ye, { className: t, ref: n, children: o });
}), at = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r, !1), /* @__PURE__ */ s(
    b,
    {
      align: "center",
      gap: 0,
      className: t,
      ref: n,
      ...r,
      children: o
    }
  );
}), it = m((e, n) => {
  const {
    className: t,
    children: o,
    icon: r,
    onClick: a,
    onDragEnd: i,
    onDragStart: l,
    draggable: c,
    label: d,
    ...f
  } = e;
  return u(f, !1), r ? /* @__PURE__ */ s(
    E,
    {
      size: 24,
      className: t,
      ref: n,
      onClick: a,
      onDragEnd: i,
      onDragStart: l,
      draggable: c,
      "aria-label": d,
      ...f,
      children: r
    }
  ) : /* @__PURE__ */ s(
    T,
    {
      className: t,
      ref: n,
      onClick: a,
      onDragEnd: i,
      onDragStart: l,
      draggable: c,
      "aria-label": d,
      ...f,
      children: o
    }
  );
}), lt = m((e, n) => {
  const { className: t, children: o, id: r, ...a } = e;
  return u(a), /* @__PURE__ */ s(
    L,
    {
      gap: 0,
      className: t,
      ref: n,
      id: r,
      role: "listbox",
      children: o
    }
  );
}), ct = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r), /* @__PURE__ */ s(b, { className: t, ref: n, children: /* @__PURE__ */ s(b, { className: "bn-mt-suggestion-menu-item-title", children: o }) });
}), dt = m((e, n) => {
  const { className: t, isSelected: o, onClick: r, item: a, id: i, ...l } = e;
  u(l);
  const c = X(null);
  return Y(() => {
    if (!c.current || !o)
      return;
    K(
      c.current,
      c.current.closest(".bn-suggestion-menu, #ai-suggestion-menu")
    ) !== "none" && c.current.scrollIntoView({ block: "nearest" });
  }, [o]), /* @__PURE__ */ v(
    b,
    {
      gap: 0,
      className: t,
      ref: j(n, c),
      id: i,
      role: "option",
      onMouseDown: (d) => d.preventDefault(),
      onClick: r,
      "aria-selected": o || void 0,
      children: [
        a.icon && /* @__PURE__ */ s(
          b,
          {
            className: "bn-mt-suggestion-menu-item-section",
            "data-position": "left",
            children: a.icon
          }
        ),
        /* @__PURE__ */ v(L, { gap: 0, className: "bn-mt-suggestion-menu-item-body", children: [
          /* @__PURE__ */ s(C, { className: "bn-mt-suggestion-menu-item-title", children: a.title }),
          /* @__PURE__ */ s(C, { className: "bn-mt-suggestion-menu-item-subtitle", children: a.subtext })
        ] }),
        a.badge && /* @__PURE__ */ s(
          b,
          {
            "data-position": "right",
            className: "bn-mt-suggestion-menu-item-section",
            children: /* @__PURE__ */ s(Ce, { size: "xs", children: a.badge })
          }
        )
      ]
    }
  );
}), ut = m((e, n) => {
  const { className: t, children: o, ...r } = e;
  return u(r), /* @__PURE__ */ s(b, { className: t, ref: n, children: o });
}), mt = m((e, n) => {
  const { className: t, ...o } = e;
  return u(o), /* @__PURE__ */ s(q, { className: t, type: "dots", size: 16, ref: n });
}), gt = m((e, n) => {
  const { className: t, children: o, id: r, columns: a, ...i } = e;
  return u(i), /* @__PURE__ */ s(
    "div",
    {
      className: t,
      style: { gridTemplateColumns: `repeat(${a}, 1fr)` },
      ref: n,
      id: r,
      role: "grid",
      children: o
    }
  );
}), ft = m((e, n) => {
  const { className: t, children: o, columns: r, ...a } = e;
  return u(a), /* @__PURE__ */ s(
    b,
    {
      className: t,
      style: { gridColumn: `1 / ${r + 1}` },
      ref: n,
      children: /* @__PURE__ */ s(b, { className: "bn-mt-suggestion-menu-item-title", children: o })
    }
  );
}), ht = m((e, n) => {
  const { className: t, isSelected: o, onClick: r, item: a, id: i, ...l } = e;
  u(l);
  const c = X(null);
  return Y(() => {
    if (!c.current || !o)
      return;
    K(
      c.current,
      c.current.closest(".bn-grid-suggestion-menu")
    ) !== "none" && c.current.scrollIntoView({ block: "nearest" });
  }, [o]), /* @__PURE__ */ s(
    "div",
    {
      className: t,
      ref: j(n, c),
      id: i,
      role: "option",
      onClick: r,
      "aria-selected": o || void 0,
      children: a.icon
    }
  );
}), bt = m((e, n) => {
  const {
    className: t,
    children: o,
    // unused, using "dots" instead
    columns: r,
    ...a
  } = e;
  return u(a), /* @__PURE__ */ s(
    q,
    {
      className: t,
      style: { gridColumn: `1 / ${r + 1}` },
      type: "dots",
      ref: n
    }
  );
}), pt = m((e, n) => {
  const { children: t, className: o, onMouseDown: r, onClick: a, ...i } = e;
  return u(i, !1), /* @__PURE__ */ s(
    T,
    {
      className: o,
      ref: n,
      onMouseDown: r,
      onClick: a,
      ...i,
      children: t
    }
  );
}), vt = m((e, n) => {
  const {
    className: t,
    children: o,
    draggable: r,
    onDragStart: a,
    onDragEnd: i,
    style: l,
    label: c,
    ...d
  } = e;
  return u(d, !1), /* @__PURE__ */ s(
    T,
    {
      className: t,
      ref: n,
      "aria-label": c,
      draggable: r,
      onDragStart: a,
      onDragEnd: i,
      style: l,
      ...d,
      children: o
    }
  );
}), D = m(
  (e, n) => {
    const {
      className: t,
      children: o,
      onMouseEnter: r,
      onMouseLeave: a,
      variant: i,
      ...l
    } = e;
    u(l);
    const { ref: c, focused: d } = Oe(), f = De(d), h = j(n, c, f);
    return /* @__PURE__ */ s(
      Ne,
      {
        className: t,
        ref: h,
        role: "toolbar",
        onMouseEnter: r,
        onMouseLeave: a,
        gap: i === "action-toolbar" ? 2 : void 0,
        children: o
      }
    );
  }
);
var ne = {
  color: void 0,
  size: void 0,
  className: void 0,
  style: void 0,
  attr: void 0
}, _ = N.createContext && /* @__PURE__ */ N.createContext(ne), St = ["attr", "size", "title"];
function yt(e, n) {
  if (e == null) return {};
  var t = wt(e, n), o, r;
  if (Object.getOwnPropertySymbols) {
    var a = Object.getOwnPropertySymbols(e);
    for (r = 0; r < a.length; r++)
      o = a[r], !(n.indexOf(o) >= 0) && Object.prototype.propertyIsEnumerable.call(e, o) && (t[o] = e[o]);
  }
  return t;
}
function wt(e, n) {
  if (e == null) return {};
  var t = {};
  for (var o in e)
    if (Object.prototype.hasOwnProperty.call(e, o)) {
      if (n.indexOf(o) >= 0) continue;
      t[o] = e[o];
    }
  return t;
}
function F() {
  return F = Object.assign ? Object.assign.bind() : function(e) {
    for (var n = 1; n < arguments.length; n++) {
      var t = arguments[n];
      for (var o in t)
        Object.prototype.hasOwnProperty.call(t, o) && (e[o] = t[o]);
    }
    return e;
  }, F.apply(this, arguments);
}
function G(e, n) {
  var t = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(e);
    n && (o = o.filter(function(r) {
      return Object.getOwnPropertyDescriptor(e, r).enumerable;
    })), t.push.apply(t, o);
  }
  return t;
}
function M(e) {
  for (var n = 1; n < arguments.length; n++) {
    var t = arguments[n] != null ? arguments[n] : {};
    n % 2 ? G(Object(t), !0).forEach(function(o) {
      xt(e, o, t[o]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : G(Object(t)).forEach(function(o) {
      Object.defineProperty(e, o, Object.getOwnPropertyDescriptor(t, o));
    });
  }
  return e;
}
function xt(e, n, t) {
  return n = Ct(n), n in e ? Object.defineProperty(e, n, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[n] = t, e;
}
function Ct(e) {
  var n = Nt(e, "string");
  return typeof n == "symbol" ? n : n + "";
}
function Nt(e, n) {
  if (typeof e != "object" || !e) return e;
  var t = e[Symbol.toPrimitive];
  if (t !== void 0) {
    var o = t.call(e, n);
    if (typeof o != "object") return o;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (n === "string" ? String : Number)(e);
}
function oe(e) {
  return e && e.map((n, t) => /* @__PURE__ */ N.createElement(n.tag, M({
    key: t
  }, n.attr), oe(n.child)));
}
function Tt(e) {
  return (n) => /* @__PURE__ */ N.createElement(kt, F({
    attr: M({}, e.attr)
  }, n), oe(e.child));
}
function kt(e) {
  var n = (t) => {
    var {
      attr: o,
      size: r,
      title: a
    } = e, i = yt(e, St), l = r || t.size || "1em", c;
    return t.className && (c = t.className), e.className && (c = (c ? c + " " : "") + e.className), /* @__PURE__ */ N.createElement("svg", F({
      stroke: "currentColor",
      fill: "currentColor",
      strokeWidth: "0"
    }, t.attr, o, i, {
      className: c,
      style: M(M({
        color: e.color || t.color
      }, t.style), e.style),
      height: l,
      width: l,
      xmlns: "http://www.w3.org/2000/svg"
    }), a && /* @__PURE__ */ N.createElement("title", null, a), e.children);
  };
  return _ !== void 0 ? /* @__PURE__ */ N.createElement(_.Consumer, null, (t) => n(t)) : n(ne);
}
function Pt(e) {
  return Tt({ attr: { viewBox: "0 0 20 20", fill: "currentColor", "aria-hidden": "true" }, child: [{ tag: "path", attr: { fillRule: "evenodd", d: "M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z", clipRule: "evenodd" }, child: [] }] })(e);
}
const I = m((e, n) => {
  const { className: t, items: o, isDisabled: r, ...a } = e;
  u(a);
  const i = o.filter((l) => l.isSelected)[0];
  return i ? /* @__PURE__ */ v(
    p,
    {
      withinPortal: !1,
      transitionProps: {
        exitDuration: 0
      },
      disabled: r,
      middlewares: { flip: !0, shift: !0, inline: !1, size: !0 },
      children: [
        /* @__PURE__ */ s(p.Target, { children: /* @__PURE__ */ s(
          T,
          {
            onMouseDown: (l) => {
              R() && l.currentTarget.focus();
            },
            leftSection: i.icon,
            rightSection: /* @__PURE__ */ s(Pt, {}),
            size: "xs",
            variant: "subtle",
            disabled: r,
            children: i.text
          }
        ) }),
        /* @__PURE__ */ s(p.Dropdown, { className: t, ref: n, children: o.map((l) => /* @__PURE__ */ s(
          p.Item,
          {
            onClick: l.onClick,
            leftSection: l.icon,
            rightSection: l.isSelected ? /* @__PURE__ */ s(z, { size: 10, className: "bn-tick-icon" }) : (
              // Ensures space for tick even if item isn't currently selected.
              /* @__PURE__ */ s("div", { className: "bn-tick-space" })
            ),
            disabled: l.isDisabled,
            children: l.text
          },
          l.text
        )) })
      ]
    }
  ) : null;
}), Ft = {
  FormattingToolbar: {
    Root: D,
    Button: O,
    Select: I
  },
  FilePanel: {
    Root: Xe,
    Button: Ye,
    FileInput: et,
    TabPanel: tt,
    TextInput: nt
  },
  GridSuggestionMenu: {
    Root: gt,
    Item: ht,
    EmptyItem: ft,
    Loader: bt
  },
  LinkToolbar: {
    Root: D,
    Button: O,
    Select: I
  },
  SideMenu: {
    Root: at,
    Button: it
  },
  SuggestionMenu: {
    Root: lt,
    Item: dt,
    EmptyItem: ct,
    Label: ut,
    Loader: mt
  },
  TableHandle: {
    Root: vt,
    ExtendButton: pt
  },
  Generic: {
    Badge: {
      Root: ze,
      Group: Be
    },
    Form: {
      Root: (e) => /* @__PURE__ */ s("div", { children: e.children }),
      TextInput: He
    },
    Menu: {
      Root: Ke,
      Trigger: We,
      Dropdown: Ze,
      Divider: qe,
      Label: Je,
      Item: Ue,
      Button: Qe
    },
    Popover: {
      Root: ot,
      Trigger: rt,
      Content: st
    },
    Toolbar: {
      Root: D,
      Button: O,
      Select: I
    }
  },
  Comments: {
    Comment: Ae,
    Editor: _e,
    Card: Ee,
    CardSection: Le,
    ExpandSectionsPrompt: je
  }
}, Mt = (e) => {
  const { className: n, theme: t, ...o } = e, r = V(), a = me(), i = (r == null ? void 0 : r.colorSchemePreference) || a, l = Fe(
    (h) => {
      if (h && (Re(h), typeof t == "object")) {
        if ("light" in t && "dark" in t) {
          A(
            t[i === "dark" ? "dark" : "light"],
            h
          );
          return;
        }
        A(t, h);
        return;
      }
    },
    [i, t]
  ), c = Q(Te), d = typeof t == "string" ? t : i !== "no-preference" ? i : "light", f = /* @__PURE__ */ s(ge.Provider, { value: Ft, children: /* @__PURE__ */ s(
    fe,
    {
      "data-mantine-color-scheme": d,
      className: B("bn-mantine", n || ""),
      theme: typeof t == "object" ? void 0 : t,
      ...o,
      ref: l
    }
  ) });
  return c ? f : /* @__PURE__ */ s(
    ke,
    {
      withCssVariables: !1,
      getRootElement: () => {
      },
      children: f
    }
  );
}, g = [
  "#FFFFFF",
  "#EFEFEF",
  "#CFCFCF",
  "#AFAFAF",
  "#7F7F7F",
  "#3F3F3F",
  "#1F1F1F",
  "#161616",
  "#0F0F0F",
  "#000000"
], H = {
  colors: {
    editor: {
      text: g[5],
      background: g[0]
    },
    menu: {
      text: g[5],
      background: g[0]
    },
    tooltip: {
      text: g[5],
      background: g[1]
    },
    hovered: {
      text: g[5],
      background: g[1]
    },
    selected: {
      text: g[0],
      background: g[5]
    },
    disabled: {
      text: g[3],
      background: g[1]
    },
    shadow: g[2],
    border: g[1],
    sideMenu: g[2],
    highlights: se
  },
  borderRadius: 6,
  fontFamily: '"Inter", "SF Pro Display", -apple-system, BlinkMacSystemFont, "Open Sans", "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif'
}, Et = {
  colors: {
    editor: {
      text: g[2],
      background: g[6]
    },
    menu: {
      text: g[2],
      background: g[6]
    },
    tooltip: {
      text: g[2],
      background: g[7]
    },
    hovered: {
      text: g[2],
      background: g[7]
    },
    selected: {
      text: g[2],
      background: g[8]
    },
    disabled: {
      text: g[5],
      background: g[7]
    },
    shadow: g[8],
    border: g[7],
    sideMenu: g[4],
    highlights: ae
  },
  borderRadius: H.borderRadius,
  fontFamily: H.fontFamily
};
export {
  Mt as BlockNoteView,
  A as applyBlockNoteCSSVariablesFromTheme,
  Ft as components,
  Et as darkDefaultTheme,
  g as defaultColorScheme,
  H as lightDefaultTheme,
  Re as removeBlockNoteCSSVariables
};
